{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ed52c95-07b9-4410-9897-39c1360e3fb5",
   "metadata": {},
   "source": [
    "# Genome Annotation\n",
    "\n",
    "Materials for the Genome Annotation a BRAKER & TSEBRA Genome Annotation workshop by Katharina Hoff (katharina.hoff@uni-greifswald.de).\n",
    "\n",
    "Please find slides for an introductory talk on genome annotation (with BRAKER and TSEBRA) at [genome_annotation_2023.pdf](genome_annotation.pdf)\n",
    "\n",
    "In the following, we will walk through the process of genome annotation on the example of a small proportion of the *Arabidopsis thaliana* genome.\n",
    "\n",
    "## Repeat masking\n",
    "\n",
    "Repetitive sequences are a huge problem for genome annotation. Some repeats only coincidentally look like protein-coding genes, others (such as transposases) are protein-coding genes, but we usually are not interested in any of these \"repeat genes\" when trying to find protein-coding genes in a novel genome. Thus, a genome should be repeat-masked prior gene prediction. \n",
    "\n",
    "Repeat masking is a resource and time-consuming step that is out of scope for this workshop. We recommend using RepeatModeler2 ([paper](https://doi.org/10.1073/pnas.1921046117), [software](https://www.repeatmasker.org/RepeatModeler/) ) to construct a species-specific repeat library and mask the genome with RepeatMasker (ideally, you will perform these computations on a node with >70 threads, in a place with very fast storage i/o, possibly using RAM instead of actual hard drive as a temporary file storage place):\n",
    "\n",
    "```\n",
    "T=72 # you need a large number of threads and fast i/o storage\n",
    "GENOME=/opt/BRAKER/example/genome.fa\n",
    "DB=some_db_name_that_fits_to_species\n",
    "\n",
    "BuildDatabase -name ${DB} ${GENOME}\n",
    "RepeatModeler -database ${DB} -pa ${T} -LTRStruct\n",
    "RepeatMasker -pa 72 -lib ${DB}-families.fa -xsmall ${GENOME}\n",
    "```\n",
    "\n",
    "This results in a file `${GENOME}.masked`. \n",
    "\n",
    "<details>\n",
    "  <summary><b>Click to learn how to mask more rigorously when needed</b></summary>\n",
    "Depending on the kind of genome, plenty of unmasked repeats may still persist. This is generally an issue to be expected in large genomes, such as vertebrate genomes, and you will notice the problem if the count of predicted proteins is extremely high. You can try to overcome \"under-masking\" with the following steps (we are suggesting to use GNU parallel to speed up the process):\n",
    "\n",
    "```\n",
    "ln -s genome.masked.fa genome.fa\n",
    "splitMfasta.pl --minsize=25000000 ${GENOME}.masked\n",
    "\n",
    "# Running TRF\n",
    "ls genome.split.*.fa | parallel 'trf {} 2 7 7 80 10 50 500 -d -m -h'\n",
    "\n",
    "# Parsing TRF output\n",
    "# The script parseTrfOutput.py is from https://github.com/gatech-genemark/BRAKER2-exp\n",
    "ls genome.split.*.fa.2.7.7.80.10.50.500.dat | parallel 'parseTrfOutput.py {} --minCopies 1 --statistics {}.STATS > {}.raw.gff 2> {}.parsedLog'\n",
    "\n",
    "# Sorting parsed output...\"\n",
    "ls genome.split.*.fa.2.7.7.80.10.50.500.dat.raw.gff | parallel 'sort -k1,1 -k4,4n -k5,5n {} > {}.sorted 2> {}.sortLog'\n",
    "\n",
    "# Merging gff...\n",
    "FILES=genome.split.*.fa.2.7.7.80.10.50.500.dat.raw.gff.sorted\n",
    "for f in $FILES\n",
    "do\n",
    "    bedtools merge -i $f | awk 'BEGIN{OFS=\"\\t\"} {print $1,\"trf\",\"repeat\",$2+1,$3,\".\",\".\",\".\",\".\"}' > $f.merged.gff 2> $f.bedtools_merge.log\n",
    "done\n",
    "\n",
    "# Masking FASTA chunk\n",
    "ls genome.split.*.fa | parallel 'bedtools maskfasta -fi {} -bed {}.2.7.7.80.10.50.500.dat.raw.gff.sorted.merged.gff -fo {}.combined.masked -soft &> {}.bedools_mask.log'\n",
    "\n",
    "# Concatenate split genome\n",
    "cat genome.split.*.fa.combined.masked > genome.fa.combined.masked\n",
    "```\n",
    "\n",
    "The file `genome.fa.combined.masked` will be more rigorously masked.\n",
    "</details>\n",
    "\n",
    "## RNA-Seq alignment with HiSat2\n",
    "\n",
    "Spliced alignments of RNA-Seq short reads are a valuable information source for predicting protein-coding genes with high accuracy.\n",
    "\n",
    "<img src=\"et-rnaseq.png\" alt=\"Figure 3 of Lomsadze et al. (2014) illustrates the use of RNA-Seq spliced alignments for predicting genes (with GeneMark-ET).\" width=\"600\"/>\n",
    "Figure 3 of Lomsadze et al. (2014) illustrates the use of RNA-Seq spliced alignments for predicting genes (with GeneMark-ET, <a href=https://doi.org/10.1093/nar/gku55\">Image Source</a>).\n",
    "\n",
    "Executing HiSat2 is out of scope for the current session. You find a readily prepared alignment file in [/opt/BRAKER/example/RNAseq.bam](/opt/BRAKER/example/RNAseq.bam).\n",
    "\n",
    "<details>\n",
    "  <summary><b>If you want to see how such a file was prepared, click here and read.</b></summary>\n",
    "  \n",
    "We will map the *Arabidopsis thaliana* Illumina RNA-Seq reads from library SRR934391 in files [SRR934391_1.fastq.gz](/home/genomics/workshop_materials/genome_annotation/sra/SRR934391_1.fastq.gz) and [SRR934391_2.fastq.gz](/home/genomics/workshop_materials/genome_annotation/sra/SRR934391_2.fastq.gz). These are paired-end data, i.e. one file contains the forward reads while the other contains in the same order the reverse reads. The length of reads is in this case 100 nt.\n",
    "\n",
    "We will use HiSat2 ([publication](https://doi.org/10.1038/s41587-019-0201-4), [software](https://github.com/DaehwanKimLab/hisat2)) to align these reads against a chunk of the *Arabidopsis thaliana* genome contained in the file [genome.fa](genome.fa). (You can in principle use any alignment tool capable of aligning RNA-seq reads to a genome, as long as it can perform spliced alignment.)\n",
    "\n",
    "First, we need to build an index from the genome file:\n",
    "\n",
    "```\n",
    "# building the hisat2 index\n",
    "hisat2-build /opt/BRAKER/example/genome.fa genome-idx 1> hisat2-build.log 2> hisat2-build.err\n",
    "```\n",
    "\n",
    "Inspect the log files [hisat2-build.log](hisat2-build.log) and [hisat2-build.err](hisat2-build.err) for possible errors.\n",
    "\n",
    "Next, we align the RNA-seq reads against the genome. Consider to **not** do this on the Cesky Krumlov Workshop AWS resources. Performing this alignment took about 7 minutes with 70 threads. The precomputed output file is provided at `/home/genomics/workshop_materials/genome_annotation/sra/SRR934391.sam`, and we will continue to use that pre-computed file.\n",
    "\n",
    "```\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR934/SRR934391/SRR934391_1.fastq.gz\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR934/SRR934391/SRR934391_2.fastq.gz\n",
    "\n",
    "RNASEQDIR=.\n",
    "\n",
    "time hisat2 -p ${T} -q -x genome-idx -1 ${RNASEQDIR}/SRR934391_1.fastq.gz \\\n",
    "    -2 ${RNASEQDIR}/SRR934391_2.fastq.gz -S rnaseq.sam \\\n",
    "    1> hisat2-align.log 2> hisat2-align.err\n",
    "```\n",
    "\n",
    "Our goal is to extract information on spliced alignments/intron positons from the alignment output file. To achieve this, we will use a tool called bam2hints that is part of the Augustus software suite ([software](https://github.com/Gaius-Augustus/Augustus)). However, this tool requires a sorted bam-file. Therefore, we first use Samtools ([paper](https://doi.org/10.1093/bioinformatics/btp352), [software](https://github.com/samtools) ) to convert the sam file to bam format:\n",
    "\n",
    "```\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with, takes ~2 minutes with 4 threads\n",
    "\n",
    "SAMFILE=/home/genomics/workshop_materials/genome_annotation/sra/SRR934391.sam\n",
    "\n",
    "time samtools view -@${T} -bSh ${SAMFILE} -o rnaseq.bam\n",
    "\n",
    "# if you computed your own rnaseq.sam file, delete it to save space on harddrive\n",
    "if [ -f rnaseq.sam ]\n",
    "then\n",
    "    rm rnaseq.sam\n",
    "fi\n",
    "```\n",
    "\n",
    "Then, we sort that bam file (this will require a bit less than 4 GB of RAM):\n",
    "\n",
    "```\n",
    "T=8 # adjust to number of threads that you booted with, takes ~2 minutes with 4 threads\n",
    "\n",
    "time samtools sort -@${T} -n rnaseq.bam -o rnaseq.s.bam\n",
    "\n",
    "# remove the unsorted bam file to save space\n",
    "rm rnaseq.bam\n",
    "```\n",
    "\n",
    "Careful, above bam file is just an demo example! We will be using a different bam file for running BRAKER because the above BAM file does not contain sufficient data for running BRAKER3, successfully!\n",
    "</details>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfcf2da9",
   "metadata": {},
   "source": [
    "## Annotation of protein coding genes\n",
    "\n",
    "Structural genome annotation is ideally performed by a combination of a statistical model (e.g. Hidden Markov Model derivate) and extrinsic evidence (e.g. from transcriptomics or known protein sequences). The statistical model parameters have to be adapted to the genomic properties of novel species. For adapting parameters, an initial set of high-quality training genes from the target species is required. This is tricky to obtain. BRAKER is a perl script that comprises several pipelines to automated the solution of this problem: fully automatically generate an initial set of training genes, train gene finders, and then predict genes with the trained parameters and extrinsic evidence.\n",
    "\n",
    "We will first take an approach to structural genome annotation that takes advantage both of RNA-Seq data, and a large database of known proteins, using BRAKER3 ([poster from PAG2023](https://www.researchgate.net/profile/Lars-Gabriel-3/publication/367409816_The_BRAKER3_Genome_Annotation_Pipeline/links/63d14cbae922c50e99c29c7a/The-BRAKER3-Genome-Annotation-Pipeline.pdf), [software](https://github.com/Gaius-Augustus/BRAKER)). If sufficient transcriptome data is available, then BRAKER3 is usually the best choice of pipeline. However, in the lack of transcriptome data, we need to consider alternative approaches. \n",
    "\n",
    "If transcriptome evidence is available but it just was not sufficient for obtaining good results with BRAKER3, then running BRAKER1 with RNA-Seq evidence ([paper](https://doi.org/10.1093/bioinformatics/btv661)) and protein supported gene prediction with BRAKER2 ([paper](https://doi.org/10.1093/nargab/lqaa108)), combined by TSEBRA ([paper](https://doi.org/10.1186/s12859-021-04482-0), [software](https://github.com/Gaius-Augustus/TSEBRA)) is often a good option.\n",
    "\n",
    "In the total absence of transcriptome data, we recommend running either BRAKER2 with a large database of proteins, alone, or for larger genomes (such as vertebrates), we the recommend application of GALBA ([preprint](https://www.biorxiv.org/content/10.1101/2023.04.10.536199v1.abstract), [software](https://github.com/Gaius-Augustus/GALBA)) with reference proteomes of a few closely related, already annotated species.\n",
    "\n",
    "### BRAKER3\n",
    "\n",
    "BRAKER3 uses spliced aligned RNA-Seq data from Hisat2 ([paper](https://www.nature.com/articles/s41587-019-0201-4), [software](https://github.com/DaehwanKimLab/hisat2)) for genome-guided transcriptome assembly with Stringtie ([paper](https://www.nature.com/articles/nbt.3122), [software](https://github.com/gpertea/stringtie)). GeneMarkS-T ([paper](https://academic.oup.com/nar/article/43/12/e78/2902598), [software](http://exon.gatech.edu/genemark/license_download.cgi)) is used to call protein coding genes in the transcripts. These transcripts are \"noisy\", therefore, a large database of proteins (e.g. an OrthoDB partition) is used to filter these predictions, using among other GeneMark-specific scripts including ProtHint ([software](https://github.com/gatech-genemark/ProtHint)) with Spaln ([paper](https://academic.oup.com/bioinformatics/article/24/21/2438/191484), [software](https://github.com/ogotoh/spaln)) the fast search tool DIAMOND ([paper](https://www.nature.com/articles/s41592-021-01101-x), [software](https://github.com/bbuchfink/diamond)). Both, the transcriptome and protein evidence is then used by GeneMark-ETP ([preprint](https://www.biorxiv.org/content/10.1101/2023.01.13.524024v1), [software](https://github.com/gatech-genemark/GeneMark-ETP)) for self-training this HMM-based gene finder. This generates a training set for AUGUSTUS ([paper](https://doi.org/10.1093/bioinformatics/btn013), [software](https://github.com/Gaius-Augustus/Augustus)). Both, the GeneMark-ETP, and the AUGUSTUS gene set incorporate the evidence to some extent, and these gene sets are merged with TSEBRA by BRAKER3.\n",
    "\n",
    "Training AUGUSTUS for a novel species usually comprises a step called etraining that adapts species-specific parameters of the statistical model of AUGUSTUS, and a step called optimize_augustus.pl that optimizes meta-parameters of that model. optimize_augustus.pl is very time-consuming, it yields usually ~2 percent points of accuracy on gene level. For this session, will disable this step with --skipOptimize. If you ever want to annotate a real new genome, make sure to delete `--skipOptimize` from your BRAKER calls (and expect substantially longer runtime). Also, GeneMark-ETP usually has a longer runtime. We will here set the maximal intergenic region for GeneMark-ETP to 10000. Please never apply this setting to a real genome annotation task, and expect a larger runtime.\n",
    "\n",
    "Note: we are in this Workshop using a toy data set instead of a real OrthoDB partition. OrthoDB partitions for real application use cases are available at https://bioinf.uni-greifswald.de/bioinf/partitioned_odb11/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a64aae87",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Tue Nov 21 10:42:10 2023: Creating directory /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER3.\n",
      "# Tue Nov 21 10:42:10 2023:Both protein and RNA-Seq data in input detected. BRAKER will be executed in ETP mode (BRAKER3).\n",
      "#*********\n",
      "# Tue Nov 21 10:42:10 2023: Log information is stored in file /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER3/braker.log\n",
      "#*********\n",
      "# WARNING: Number of reliable training genes is low (114). Recommended are at least 600 genes\n",
      "#*********\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t9m45.779s\n",
      "user\t18m20.790s\n",
      "sys\t3m48.617s\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "T=6 # adjust to number of threads that you booted with, takes ~30 minutes with 4 threads\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d BRAKER3 ]\n",
    "then\n",
    "    rm -rf BRAKER3\n",
    "fi\n",
    "\n",
    "ORTHODB=/opt/BRAKER/example/proteins.fa # adjust to suitable clade of real OrthoDB from https://bioinf.uni-greifswald.de/bioinf/partitioned_odb11/\n",
    "\n",
    "# run BRAKER3\n",
    "time braker.pl --workingdir=BRAKER3 --genome=/opt/BRAKER/example/genome.fa \\\n",
    "    --bam=/opt/BRAKER/example/RNAseq.bam \\\n",
    "    --prot_seq=${ORTHODB} --AUGUSTUS_BIN_PATH=/usr/bin/ \\\n",
    "    --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --gm_max_intergenic 10000 --skipOptimize # remember to remove both these options for real jobs!\n",
    "    # this call takes a few minutes even with --skipOptimize and --gm_max_intergenic 10000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37295089",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Out of time, job died? Click here.</b></summary>\n",
    "If you ran out of time (the BRAKER3 job takes substantial time), you may copy the most important files as follows from a notebook cell:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "cp -r BRAKER3_precomputed_results BRAKER3\n",
    "```\n",
    "</details>\n",
    "\n",
    "Let's inspect the output, the most important files are braker.gtf, Augustus/augustus.hints.gtf, and GeneMark-ETP/genemark.gtf:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4444062",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 jovyan root 377K Nov 21 10:51 Augustus/augustus.hints.gtf\n",
      "-rw-r--r-- 1 jovyan root 393K Nov 21 10:51 braker.gtf\n",
      "-rw-r--r-- 1 jovyan root 402K Nov 21 10:51 GeneMark-ETP/genemark.gtf\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "cd BRAKER3\n",
    "ls -lh braker.gtf Augustus/augustus.hints.gtf GeneMark-ETP/genemark.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be4ff890",
   "metadata": {},
   "source": [
    "The file [BRAKER3/what-to-cite.txt](BRAKER3/what-to-cite.txt) advises you on what papers should be cited if you were going to publish a manuscript on a gene set produced with BRAKER3.\n",
    "\n",
    "Usually, the braker.gtf is the main output. However, because of the way that TSEBRA combines the augustus.hints.gtf and the genemark.gtf file with the hintsfile.gff, this may not be the optimal gene set. We can investigate this by descriptive statistics of gene sets (e.g. how many genes are in a gene set, how many transcripts, how many exons does an average transcript have?), or we measure sensivity according to a conserved gene set, e.g. BUSCO or OMArk. Here, we will use BUSCO. Before running BUSCO, we need to make sure that we have protein sequences of all three gene sets (only the braker.aa exists by default):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "625a0b68",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 jovyan root 121K Nov 21 10:51 Augustus/augustus.hints.aa\n",
      "-rw-r--r-- 1 jovyan root 124K Nov 21 10:51 braker.aa\n",
      "-rw-r--r-- 1 jovyan root 115K Nov 21 10:56 GeneMark-ETP/genemark.aa\n",
      "Counting number of protein sequences = transcripts\n",
      "braker.aa:289\n",
      "GeneMark-ETP/genemark.aa:270\n",
      "Augustus/augustus.hints.aa:281\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "# generate protein (and coding seq file) from AUGUSTUS predictions\n",
    "cd BRAKER3/Augustus\n",
    "# getAnnoFastaFromJoingenes.py -g /opt/BRAKER/example/genome.fa -o augustus.hints -f augustus.hints.gtf\n",
    "# generate protein (and coding seq file) from GeneMark-ETP predictions\n",
    "cd ../GeneMark-ETP\n",
    "getAnnoFastaFromJoingenes.py -g /opt/BRAKER/example/genome.fa -o genemark -f genemark.gtf\n",
    "# see file sizes\n",
    "cd ../\n",
    "ls -lh braker.aa GeneMark-ETP/genemark.aa Augustus/augustus.hints.aa\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" braker.aa GeneMark-ETP/genemark.aa Augustus/augustus.hints.aa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afee189b",
   "metadata": {},
   "source": [
    "GALBA has a simple script to compute the ratio of mono- to multi-exonic genes (only counting one isoform if one gene has several alternative isoforms, that's why the transcript number differs from the number above for methods that contain alternative transcripts, such as AUGUSTUS and BRAKER):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2880a8be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing some descriptive statistics for BRAKER:\n",
      "Number of transcripts: 281\n",
      "Largest number of exons in all transcripts: 30\n",
      "Monoexonic transcripts: 72\n",
      "Multiexonic transcripts: 209\n",
      "Mono:Mult Ratio: 0.34\n",
      "Boxplot of number of exons per transcript:\n",
      "Min: 1\n",
      "25%: 1\n",
      "50%: 4\n",
      "75%: 8\n",
      "Max: 30\n",
      "\n",
      "Doing the same for Augustus:\n",
      "Number of transcripts: 271\n",
      "Largest number of exons in all transcripts: 30\n",
      "Monoexonic transcripts: 68\n",
      "Multiexonic transcripts: 203\n",
      "Mono:Mult Ratio: 0.33\n",
      "Boxplot of number of exons per transcript:\n",
      "Min: 1\n",
      "25%: 1\n",
      "50%: 4\n",
      "75%: 7\n",
      "Max: 30\n",
      "\n",
      "And for GeneMark-ETP:\n",
      "Number of transcripts: 270\n",
      "Largest number of exons in all transcripts: 30\n",
      "Monoexonic transcripts: 59\n",
      "Multiexonic transcripts: 211\n",
      "Mono:Mult Ratio: 0.28\n",
      "Boxplot of number of exons per transcript:\n",
      "Min: 1\n",
      "25%: 2\n",
      "50%: 4\n",
      "75%: 8\n",
      "Max: 30\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "cd BRAKER3\n",
    "echo \"Computing some descriptive statistics for BRAKER:\"\n",
    "analyze_exons.py -f braker.gtf\n",
    "echo \"\"\n",
    "echo \"Doing the same for Augustus:\"\n",
    "analyze_exons.py -f Augustus/augustus.hints.gtf\n",
    "echo \"\"\n",
    "echo \"And for GeneMark-ETP:\"\n",
    "analyze_exons.py -f GeneMark-ETP/genemark.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47e1cedd",
   "metadata": {},
   "source": [
    "#### BUSCO assessment\n",
    "\n",
    "BUSCO ([paper](https://doi.org/10.1002/cpz1.323), [software](https://gitlab.com/ezlab/busco)) can provide information on sensitivity with respect to a clade-specific core gene set. We will in the following use BUSCO to compare sensitivity in the BRAKER3, AUGUSTUS, and GeneMark-ETP gene set.\n",
    "\n",
    "First, we find the closest BUSCO lineage (we are working on *Arabidopsis thaliana*):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f6117f1f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%script bash\n",
    "\n",
    "# BUSCO resides in its own conda environment in our container, that's why we need to activate it.\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "busco --list-datasets > busco_lineages.txt 2> busco_lineages.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "885e452c",
   "metadata": {},
   "source": [
    "All available lineages are now in [busco_lineages.txt](busco_lineages.txt). (Check [busco_lineages.log](busco_lineages.log) for possible errors.)\n",
    "\n",
    "Check at [NCBI taxonomy](https://www.ncbi.nlm.nih.gov/taxonomy) the lineage of the target *Arabidopsis*. I believe the lineage is:\n",
    "\n",
    "`cellular organisms; Eukaryota; Viridiplantae; Streptophyta; Streptophytina; Embryophyta; Tracheophyta; Euphyllophyta; Spermatophyta; Magnoliopsida; Mesangiospermae; eudicotyledons; Gunneridae; Pentapetalae; rosids; malvids; Brassicales; Brassicaceae; Camelineae`\n",
    "\n",
    "Now find a related lineage in [busco_lineages.txt](busco_lineages.txt). \n",
    "\n",
    "`brassicales_odb10` is the closest lineage.\n",
    "\n",
    "Next, we run a BUSCO assessment on all gene sets (this takes ~3 minutes with 4 threads):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4e868f00",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing braker...\n",
      "Processing augustus...\n",
      "Processing genemark...\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "T=6 # adjust to number of threads that you booted with\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "cd BRAKER3\n",
    "# create links if not already present\n",
    "if [ ! -L augustus.aa ]\n",
    "then\n",
    "    ln -s Augustus/augustus.hints.aa augustus.aa\n",
    "fi\n",
    "\n",
    "if [ ! -L genemark.aa ]\n",
    "then\n",
    "    ln -s GeneMark-ETP/genemark.aa genemark.aa\n",
    "    sleep 3 # not sure why we need to wait a few seconds, but otherwise system doesn't find the file\n",
    "fi\n",
    "\n",
    "GENESETS=(braker augustus genemark)\n",
    "\n",
    "# Process all gene sets from above list\n",
    "for g in ${GENESETS[@]}; do\n",
    "    echo \"Processing ${g}...\"\n",
    "    # delete old output if existing\n",
    "    if [ -d busco_${g} ]\n",
    "    then\n",
    "        rm -r busco_${g}\n",
    "    fi\n",
    "    # run BUSCO\n",
    "    busco -m proteins -i ${g}.aa -o busco_${g} \\\n",
    "        -l brassicales_odb10 -c ${T} &> busco_${g}.log\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9260cae",
   "metadata": {},
   "source": [
    "Next, we visualize the BUSCO results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "84a8a2f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%script bash\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "cd BRAKER3\n",
    "\n",
    "# create BUSCO_summaries folder if not present\n",
    "if ! [ -d BUSCO_summaries ]\n",
    "then\n",
    "    mkdir BUSCO_summaries\n",
    "fi\n",
    "\n",
    "# copy all BUSCO results into the summaries folder\n",
    "cp busco_*/short_summary*.txt BUSCO_summaries\n",
    "\n",
    "# generate BUSCO plot\n",
    "generate_plot.py -wd BUSCO_summaries &> generate_plot.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b770b22",
   "metadata": {},
   "source": [
    "Check the file [generate_plot.log](generate_plot.log) for possible errors. This results in the following figure (stored at [BRAKER3/BUSCO_summaries/busco_figure.png](BRAKER3/BUSCO_summaries/busco_figure.png), reload website in order to display image after generating it):\n",
    "\n",
    "<img src=\"BRAKER3/BUSCO_summaries/busco_figure.png\" alt=\"BUSCO results\" width=\"400\"/>\n",
    "\n",
    "The data that we used in this session was selected purely on the criterion of feasible runtime. In a real scenario, with a complete genome, the BUSCO plot should contain a much larger number of complete BUSCOs, and you are usually happy if the number BUSCOs in the BRAKER3 gene set is higher or equal to the number of BUSCOs detected in the AUGUSTUS and GeneMark-ETP set, while the total number of transcripts does not grow into an unexpected way (e.g. having 80.000 proteins in a BRAKER gene set does seem odd...).\n",
    "\n",
    "But what can we do if there is no RNA-Seq data for a particular species? In that case, we can resort to using either BRAKER2 (for small and medium sized genomes, with a large database of proteins that might be only remotely related), or we may use GALBA (for large vertebrate genomes, with a few closely related reference proteomes).\n",
    "\n",
    "### BRAKER2\n",
    "\n",
    "BRAKER2 ([paper](https://doi.org/10.1093/nargab/lqaa108)) uses spliced alignment information from a huge database of proteins against the target genome. We typically use OrthoDB partitions of clades, hosted at https://bioinf.uni-greifswald.de/bioinf/partitioned_odb11/. Note: a set of proteins from one or a few related species is not sufficient for running BRAKER2. A particular set of proteins of a closely related species can be appended to a larger database for running BRAKER2. However, BRAKER2 is not an ideal tool for recovering a complete set of proteins from a related species.\n",
    "\n",
    "The following call of BRAKER2 takes ~15 minutes on 4 threads, even when optimizing AUGUSTUS parameters is disabled:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9e55ca6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Tue Nov 21 12:29:45 2023: Log information is stored in file /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/braker.log\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ProtHint Version 2.6.0\n",
      "Copyright 2019, Georgia Institute of Technology, USA\n",
      "\n",
      "Please cite\n",
      "  - ProtHint: https://doi.org/10.1093/nargab/lqaa026\n",
      "  - DIAMOND:  https://doi.org/10.1038/nmeth.3176\n",
      "  - Spaln:    https://doi.org/10.1093/bioinformatics/btn460\n",
      "\n",
      "Called from: /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2\n",
      "Cmd: /opt/ETP/bin/gmes/ProtHint/bin/prothint.py --threads=6 --geneMarkGtf /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/GeneMark-ES/genemark.gtf /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/genome.fa /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/proteins.fa\n",
      "\n",
      "[Tue Nov 21 12:31:20 2023] Pre-processing protein input\n",
      "[Tue Nov 21 12:31:20 2023] Skipping GeneMark-ES, using the supplied gene seeds file instead\n",
      "[Tue Nov 21 12:31:20 2023] Translating gene seeds to proteins\n",
      "[Tue Nov 21 12:31:20 2023] Translation of seeds finished\n",
      "[Tue Nov 21 12:31:20 2023] Running DIAMOND\n",
      "diamond v0.9.24.125 | by Benjamin Buchfink <buchfink@gmail.com>\n",
      "Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>\n",
      "Check http://github.com/bbuchfink/diamond for updates.\n",
      "\n",
      "#CPU threads: 6\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Database file: /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/protdmmkh8b8\n",
      "Opening the database file...  [0.000186s]\n",
      "Loading sequences...  [0.020096s]\n",
      "Masking sequences...  [0.129284s]\n",
      "Writing sequences...  [0.006225s]\n",
      "Hashing sequences...  [0.002151s]\n",
      "Loading sequences...  [1.9e-05s]\n",
      "Writing trailer...  [0.000162s]\n",
      "Closing the input file...  [2.1e-05s]\n",
      "Closing the database file...  [2.7e-05s]\n",
      "Database hash = e1048f69f134493ac567793c01fff94c\n",
      "Processed 5733 sequences, 2599346 letters.\n",
      "Total time = 0.158344s\n",
      "diamond v0.9.24.125 | by Benjamin Buchfink <buchfink@gmail.com>\n",
      "Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>\n",
      "Check http://github.com/bbuchfink/diamond for updates.\n",
      "\n",
      "#CPU threads: 6\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: \n",
      "Opening the database...  [5.7e-05s]\n",
      "#Target sequences to report alignments for: 25\n",
      "Opening the input file...  [2.7e-05s]\n",
      "Opening the output file...  [2.2e-05s]\n",
      "Loading query sequences...  [0.001082s]\n",
      "Masking queries...  [0.015855s]\n",
      "Building query seed set...  [0.004226s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.001539s]\n",
      "Allocating buffers...  [5.4e-05s]\n",
      "Loading reference sequences...  [0.010124s]\n",
      "Building reference histograms...  [0.023347s]\n",
      "Allocating buffers...  [5.9e-05s]\n",
      "Initializing temporary storage...  [0.003351s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 0.\n",
      "Building reference index...  [0.031341s]\n",
      "Building query index...  [0.002232s]\n",
      "Building seed filter...  [0.002639s]\n",
      "Searching alignments...  [0.026775s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 1.\n",
      "Building reference index...  [0.03387s]\n",
      "Building query index...  [0.002515s]\n",
      "Building seed filter...  [0.003225s]\n",
      "Searching alignments...  [0.017733s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 2.\n",
      "Building reference index...  [0.035331s]\n",
      "Building query index...  [0.002801s]\n",
      "Building seed filter...  [0.002714s]\n",
      "Searching alignments...  [0.018433s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 3.\n",
      "Building reference index...  [0.030125s]\n",
      "Building query index...  [0.002376s]\n",
      "Building seed filter...  [0.003411s]\n",
      "Searching alignments...  [0.021538s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 0.\n",
      "Building reference index...  [0.029745s]\n",
      "Building query index...  [0.003112s]\n",
      "Building seed filter...  [0.00303s]\n",
      "Searching alignments...  [0.019954s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 1.\n",
      "Building reference index...  [0.032338s]\n",
      "Building query index...  [0.002581s]\n",
      "Building seed filter...  [0.002723s]\n",
      "Searching alignments...  [0.016595s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 2.\n",
      "Building reference index...  [0.038012s]\n",
      "Building query index...  [0.002636s]\n",
      "Building seed filter...  [0.002916s]\n",
      "Searching alignments...  [0.021474s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 3.\n",
      "Building reference index...  [0.03728s]\n",
      "Building query index...  [0.002506s]\n",
      "Building seed filter...  [0.003161s]\n",
      "Searching alignments...  [0.020171s]\n",
      "Deallocating buffers...  [0.000779s]\n",
      "Computing alignments...  [0.186452s]\n",
      "Deallocating reference...  [0.000489s]\n",
      "Loading reference sequences...  [2.9e-05s]\n",
      "Deallocating buffers...  [6e-06s]\n",
      "Deallocating queries...  [6.8e-05s]\n",
      "Loading query sequences...  [2.5e-05s]\n",
      "Closing the input file...  [2.1e-05s]\n",
      "Closing the output file...  [6.7e-05s]\n",
      "Closing the database file...  [1.5e-05s]\n",
      "Deallocating taxonomy...  [5e-06s]\n",
      "Total time = 0.724857s\n",
      "Reported 5830 pairwise alignments, 5830 HSPs.\n",
      "239 queries aligned.\n",
      "[Tue Nov 21 12:31:21 2023] DIAMOND finished\n",
      "[Tue Nov 21 12:31:21 2023] Preparing pairs for alignments\n",
      "[Tue Nov 21 12:31:21 2023] Preparation of pairs finished\n",
      "[Tue Nov 21 12:31:21 2023] Starting spliced alignment with spaln\n",
      "[Tue Nov 21 12:31:21 2023] Loading alignment pairs into memory\n",
      "[Tue Nov 21 12:31:21 2023] Pairs loaded. Number of pairs to align: 5830\n",
      "[Tue Nov 21 12:31:21 2023] Starting the alignments\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 0/5830 (0.0%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 6/5830 (0.1%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 12/5830 (0.2%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 18/5830 (0.3%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 24/5830 (0.4%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 30/5830 (0.5%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 35/5830 (0.6%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 41/5830 (0.7%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 47/5830 (0.8%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 53/5830 (0.9%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 59/5830 (1.0%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 65/5830 (1.1%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 70/5830 (1.2%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 76/5830 (1.3%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 82/5830 (1.4%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 88/5830 (1.5%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 94/5830 (1.6%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 100/5830 (1.7%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 105/5830 (1.8%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 111/5830 (1.9%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 117/5830 (2.0%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 123/5830 (2.1%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 129/5830 (2.2%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 135/5830 (2.3%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 140/5830 (2.4%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 146/5830 (2.5%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 152/5830 (2.6%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 158/5830 (2.7%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 164/5830 (2.8%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 170/5830 (2.9%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 175/5830 (3.0%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 181/5830 (3.1%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 187/5830 (3.2%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 193/5830 (3.3%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 199/5830 (3.4%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 205/5830 (3.5%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 210/5830 (3.6%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 216/5830 (3.7%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 222/5830 (3.8%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 228/5830 (3.9%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 234/5830 (4.0%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 240/5830 (4.1%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 245/5830 (4.2%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 251/5830 (4.3%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 257/5830 (4.4%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 263/5830 (4.5%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 269/5830 (4.6%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 275/5830 (4.7%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 280/5830 (4.8%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 286/5830 (4.9%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 292/5830 (5.0%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 298/5830 (5.1%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 304/5830 (5.2%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 309/5830 (5.3%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 315/5830 (5.4%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 321/5830 (5.5%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 327/5830 (5.6%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 333/5830 (5.7%)\n",
      "[Tue Nov 21 12:31:21 2023] Enqueueing pair 339/5830 (5.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 344/5830 (5.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 350/5830 (6.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 356/5830 (6.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 362/5830 (6.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 368/5830 (6.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 374/5830 (6.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 379/5830 (6.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 385/5830 (6.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 391/5830 (6.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 397/5830 (6.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 403/5830 (6.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 409/5830 (7.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 414/5830 (7.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 420/5830 (7.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 426/5830 (7.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 432/5830 (7.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 438/5830 (7.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 444/5830 (7.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 449/5830 (7.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 455/5830 (7.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 461/5830 (7.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 467/5830 (8.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 473/5830 (8.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 479/5830 (8.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 484/5830 (8.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 490/5830 (8.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 496/5830 (8.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 502/5830 (8.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 508/5830 (8.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 514/5830 (8.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 519/5830 (8.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 525/5830 (9.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 531/5830 (9.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 537/5830 (9.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 543/5830 (9.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 549/5830 (9.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 554/5830 (9.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 560/5830 (9.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 566/5830 (9.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 572/5830 (9.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 578/5830 (9.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 583/5830 (10.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 589/5830 (10.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 595/5830 (10.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 601/5830 (10.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 607/5830 (10.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 613/5830 (10.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 618/5830 (10.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 624/5830 (10.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 630/5830 (10.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 636/5830 (10.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 642/5830 (11.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 648/5830 (11.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 653/5830 (11.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 659/5830 (11.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 665/5830 (11.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 671/5830 (11.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 677/5830 (11.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 683/5830 (11.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 688/5830 (11.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 694/5830 (11.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 700/5830 (12.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 706/5830 (12.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 712/5830 (12.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 718/5830 (12.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 723/5830 (12.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 729/5830 (12.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 735/5830 (12.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 741/5830 (12.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 747/5830 (12.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 753/5830 (12.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 758/5830 (13.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 764/5830 (13.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 770/5830 (13.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 776/5830 (13.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 782/5830 (13.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 788/5830 (13.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 793/5830 (13.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 799/5830 (13.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 805/5830 (13.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 811/5830 (13.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 817/5830 (14.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 823/5830 (14.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 828/5830 (14.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 834/5830 (14.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 840/5830 (14.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 846/5830 (14.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 852/5830 (14.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 858/5830 (14.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 863/5830 (14.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 869/5830 (14.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 875/5830 (15.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 881/5830 (15.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 887/5830 (15.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 892/5830 (15.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 898/5830 (15.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 904/5830 (15.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 910/5830 (15.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 916/5830 (15.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 922/5830 (15.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 927/5830 (15.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 933/5830 (16.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 939/5830 (16.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 945/5830 (16.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 951/5830 (16.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 957/5830 (16.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 962/5830 (16.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 968/5830 (16.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 974/5830 (16.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 980/5830 (16.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 986/5830 (16.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 992/5830 (17.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 997/5830 (17.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1003/5830 (17.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1009/5830 (17.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1015/5830 (17.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1021/5830 (17.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1027/5830 (17.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1032/5830 (17.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1038/5830 (17.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1044/5830 (17.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1050/5830 (18.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1056/5830 (18.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1062/5830 (18.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1067/5830 (18.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1073/5830 (18.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1079/5830 (18.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1085/5830 (18.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1091/5830 (18.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1097/5830 (18.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1102/5830 (18.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1108/5830 (19.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1114/5830 (19.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1120/5830 (19.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1126/5830 (19.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1132/5830 (19.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1137/5830 (19.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1143/5830 (19.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1149/5830 (19.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1155/5830 (19.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1161/5830 (19.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1166/5830 (20.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1172/5830 (20.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1178/5830 (20.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1184/5830 (20.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1190/5830 (20.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1196/5830 (20.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1201/5830 (20.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1207/5830 (20.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1213/5830 (20.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1219/5830 (20.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1225/5830 (21.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1231/5830 (21.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1236/5830 (21.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1242/5830 (21.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1248/5830 (21.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1254/5830 (21.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1260/5830 (21.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1266/5830 (21.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1271/5830 (21.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1277/5830 (21.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1283/5830 (22.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1289/5830 (22.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1295/5830 (22.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1301/5830 (22.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1306/5830 (22.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1312/5830 (22.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1318/5830 (22.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1324/5830 (22.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1330/5830 (22.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1336/5830 (22.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1341/5830 (23.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1347/5830 (23.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1353/5830 (23.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1359/5830 (23.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1365/5830 (23.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1371/5830 (23.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1376/5830 (23.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1382/5830 (23.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1388/5830 (23.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1394/5830 (23.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1400/5830 (24.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1406/5830 (24.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1411/5830 (24.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1417/5830 (24.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1423/5830 (24.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1429/5830 (24.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1435/5830 (24.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1441/5830 (24.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1446/5830 (24.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1452/5830 (24.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1458/5830 (25.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1464/5830 (25.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1470/5830 (25.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1475/5830 (25.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1481/5830 (25.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1487/5830 (25.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1493/5830 (25.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1499/5830 (25.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1505/5830 (25.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1510/5830 (25.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1516/5830 (26.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1522/5830 (26.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1528/5830 (26.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1534/5830 (26.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1540/5830 (26.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1545/5830 (26.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1551/5830 (26.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1557/5830 (26.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1563/5830 (26.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1569/5830 (26.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1575/5830 (27.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1580/5830 (27.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1586/5830 (27.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1592/5830 (27.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1598/5830 (27.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1604/5830 (27.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1610/5830 (27.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1615/5830 (27.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1621/5830 (27.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1627/5830 (27.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1633/5830 (28.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1639/5830 (28.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1645/5830 (28.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1650/5830 (28.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1656/5830 (28.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1662/5830 (28.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1668/5830 (28.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1674/5830 (28.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1680/5830 (28.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1685/5830 (28.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1691/5830 (29.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1697/5830 (29.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1703/5830 (29.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1709/5830 (29.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1715/5830 (29.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1720/5830 (29.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1726/5830 (29.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1732/5830 (29.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1738/5830 (29.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1744/5830 (29.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1749/5830 (30.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1755/5830 (30.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1761/5830 (30.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1767/5830 (30.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1773/5830 (30.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1779/5830 (30.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1784/5830 (30.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1790/5830 (30.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1796/5830 (30.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1802/5830 (30.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1808/5830 (31.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1814/5830 (31.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1819/5830 (31.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1825/5830 (31.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1831/5830 (31.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1837/5830 (31.5%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1843/5830 (31.6%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1849/5830 (31.7%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1854/5830 (31.8%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1860/5830 (31.9%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1866/5830 (32.0%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1872/5830 (32.1%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1878/5830 (32.2%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1884/5830 (32.3%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1889/5830 (32.4%)\n",
      "[Tue Nov 21 12:31:22 2023] Enqueueing pair 1895/5830 (32.5%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1901/5830 (32.6%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1907/5830 (32.7%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1913/5830 (32.8%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1919/5830 (32.9%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1924/5830 (33.0%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1930/5830 (33.1%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1936/5830 (33.2%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1942/5830 (33.3%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1948/5830 (33.4%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1954/5830 (33.5%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1959/5830 (33.6%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1965/5830 (33.7%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1971/5830 (33.8%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1977/5830 (33.9%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1983/5830 (34.0%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1989/5830 (34.1%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 1994/5830 (34.2%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2000/5830 (34.3%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2006/5830 (34.4%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2012/5830 (34.5%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2018/5830 (34.6%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2024/5830 (34.7%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2029/5830 (34.8%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2035/5830 (34.9%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2041/5830 (35.0%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2047/5830 (35.1%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2053/5830 (35.2%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2058/5830 (35.3%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2064/5830 (35.4%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2070/5830 (35.5%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2076/5830 (35.6%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2082/5830 (35.7%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2088/5830 (35.8%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2093/5830 (35.9%)\n",
      "[Tue Nov 21 12:31:26 2023] Enqueueing pair 2099/5830 (36.0%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2105/5830 (36.1%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2111/5830 (36.2%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2117/5830 (36.3%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2123/5830 (36.4%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2128/5830 (36.5%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2134/5830 (36.6%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2140/5830 (36.7%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2146/5830 (36.8%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2152/5830 (36.9%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2158/5830 (37.0%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2163/5830 (37.1%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2169/5830 (37.2%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2175/5830 (37.3%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2181/5830 (37.4%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2187/5830 (37.5%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2193/5830 (37.6%)\n",
      "[Tue Nov 21 12:31:27 2023] Enqueueing pair 2198/5830 (37.7%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2204/5830 (37.8%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2210/5830 (37.9%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2216/5830 (38.0%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2222/5830 (38.1%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2228/5830 (38.2%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2233/5830 (38.3%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2239/5830 (38.4%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2245/5830 (38.5%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2251/5830 (38.6%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2257/5830 (38.7%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2263/5830 (38.8%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2268/5830 (38.9%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2274/5830 (39.0%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2280/5830 (39.1%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2286/5830 (39.2%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2292/5830 (39.3%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2298/5830 (39.4%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2303/5830 (39.5%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2309/5830 (39.6%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2315/5830 (39.7%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2321/5830 (39.8%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2327/5830 (39.9%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2332/5830 (40.0%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2338/5830 (40.1%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2344/5830 (40.2%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2350/5830 (40.3%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2356/5830 (40.4%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2362/5830 (40.5%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2367/5830 (40.6%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2373/5830 (40.7%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2379/5830 (40.8%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2385/5830 (40.9%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2391/5830 (41.0%)\n",
      "[Tue Nov 21 12:31:28 2023] Enqueueing pair 2397/5830 (41.1%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2402/5830 (41.2%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2408/5830 (41.3%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2414/5830 (41.4%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2420/5830 (41.5%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2426/5830 (41.6%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2432/5830 (41.7%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2437/5830 (41.8%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2443/5830 (41.9%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2449/5830 (42.0%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2455/5830 (42.1%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2461/5830 (42.2%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2467/5830 (42.3%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2472/5830 (42.4%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2478/5830 (42.5%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2484/5830 (42.6%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2490/5830 (42.7%)\n",
      "[Tue Nov 21 12:31:29 2023] Enqueueing pair 2496/5830 (42.8%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2502/5830 (42.9%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2507/5830 (43.0%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2513/5830 (43.1%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2519/5830 (43.2%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2525/5830 (43.3%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2531/5830 (43.4%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2537/5830 (43.5%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2542/5830 (43.6%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2548/5830 (43.7%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2554/5830 (43.8%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2560/5830 (43.9%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2566/5830 (44.0%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2572/5830 (44.1%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2577/5830 (44.2%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2583/5830 (44.3%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2589/5830 (44.4%)\n",
      "[Tue Nov 21 12:31:32 2023] Enqueueing pair 2595/5830 (44.5%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2601/5830 (44.6%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2607/5830 (44.7%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2612/5830 (44.8%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2618/5830 (44.9%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2624/5830 (45.0%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2630/5830 (45.1%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2636/5830 (45.2%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2641/5830 (45.3%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2647/5830 (45.4%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2653/5830 (45.5%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2659/5830 (45.6%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2665/5830 (45.7%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2671/5830 (45.8%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2676/5830 (45.9%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2682/5830 (46.0%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2688/5830 (46.1%)\n",
      "[Tue Nov 21 12:31:33 2023] Enqueueing pair 2694/5830 (46.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2700/5830 (46.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2706/5830 (46.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2711/5830 (46.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2717/5830 (46.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2723/5830 (46.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2729/5830 (46.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2735/5830 (46.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2741/5830 (47.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2746/5830 (47.1%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2752/5830 (47.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2758/5830 (47.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2764/5830 (47.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2770/5830 (47.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2776/5830 (47.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2781/5830 (47.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2787/5830 (47.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2793/5830 (47.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2799/5830 (48.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2805/5830 (48.1%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2811/5830 (48.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2816/5830 (48.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2822/5830 (48.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2828/5830 (48.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2834/5830 (48.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2840/5830 (48.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2846/5830 (48.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2851/5830 (48.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2857/5830 (49.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2863/5830 (49.1%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2869/5830 (49.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2875/5830 (49.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2881/5830 (49.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2886/5830 (49.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2892/5830 (49.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2898/5830 (49.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2904/5830 (49.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2910/5830 (49.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2915/5830 (50.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2921/5830 (50.1%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2927/5830 (50.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2933/5830 (50.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2939/5830 (50.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2945/5830 (50.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2950/5830 (50.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2956/5830 (50.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2962/5830 (50.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2968/5830 (50.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2974/5830 (51.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2980/5830 (51.1%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2985/5830 (51.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2991/5830 (51.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 2997/5830 (51.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3003/5830 (51.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3009/5830 (51.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3015/5830 (51.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3020/5830 (51.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3026/5830 (51.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3032/5830 (52.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3038/5830 (52.1%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3044/5830 (52.2%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3050/5830 (52.3%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3055/5830 (52.4%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3061/5830 (52.5%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3067/5830 (52.6%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3073/5830 (52.7%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3079/5830 (52.8%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3085/5830 (52.9%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3090/5830 (53.0%)\n",
      "[Tue Nov 21 12:31:34 2023] Enqueueing pair 3096/5830 (53.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3102/5830 (53.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3108/5830 (53.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3114/5830 (53.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3120/5830 (53.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3125/5830 (53.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3131/5830 (53.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3137/5830 (53.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3143/5830 (53.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3149/5830 (54.0%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3155/5830 (54.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3160/5830 (54.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3166/5830 (54.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3172/5830 (54.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3178/5830 (54.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3184/5830 (54.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3190/5830 (54.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3195/5830 (54.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3201/5830 (54.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3207/5830 (55.0%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3213/5830 (55.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3219/5830 (55.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3224/5830 (55.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3230/5830 (55.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3236/5830 (55.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3242/5830 (55.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3248/5830 (55.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3254/5830 (55.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3259/5830 (55.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3265/5830 (56.0%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3271/5830 (56.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3277/5830 (56.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3283/5830 (56.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3289/5830 (56.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3294/5830 (56.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3300/5830 (56.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3306/5830 (56.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3312/5830 (56.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3318/5830 (56.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3324/5830 (57.0%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3329/5830 (57.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3335/5830 (57.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3341/5830 (57.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3347/5830 (57.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3353/5830 (57.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3359/5830 (57.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3364/5830 (57.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3370/5830 (57.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3376/5830 (57.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3382/5830 (58.0%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3388/5830 (58.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3394/5830 (58.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3399/5830 (58.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3405/5830 (58.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3411/5830 (58.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3417/5830 (58.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3423/5830 (58.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3429/5830 (58.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3434/5830 (58.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3440/5830 (59.0%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3446/5830 (59.1%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3452/5830 (59.2%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3458/5830 (59.3%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3464/5830 (59.4%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3469/5830 (59.5%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3475/5830 (59.6%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3481/5830 (59.7%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3487/5830 (59.8%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3493/5830 (59.9%)\n",
      "[Tue Nov 21 12:31:39 2023] Enqueueing pair 3498/5830 (60.0%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3504/5830 (60.1%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3510/5830 (60.2%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3516/5830 (60.3%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3522/5830 (60.4%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3528/5830 (60.5%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3533/5830 (60.6%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3539/5830 (60.7%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3545/5830 (60.8%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3551/5830 (60.9%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3557/5830 (61.0%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3563/5830 (61.1%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3568/5830 (61.2%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3574/5830 (61.3%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3580/5830 (61.4%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3586/5830 (61.5%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3592/5830 (61.6%)\n",
      "[Tue Nov 21 12:31:40 2023] Enqueueing pair 3598/5830 (61.7%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3603/5830 (61.8%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3609/5830 (61.9%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3615/5830 (62.0%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3621/5830 (62.1%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3627/5830 (62.2%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3633/5830 (62.3%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3638/5830 (62.4%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3644/5830 (62.5%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3650/5830 (62.6%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3656/5830 (62.7%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3662/5830 (62.8%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3668/5830 (62.9%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3673/5830 (63.0%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3679/5830 (63.1%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3685/5830 (63.2%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3691/5830 (63.3%)\n",
      "[Tue Nov 21 12:31:42 2023] Enqueueing pair 3697/5830 (63.4%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3703/5830 (63.5%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3708/5830 (63.6%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3714/5830 (63.7%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3720/5830 (63.8%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3726/5830 (63.9%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3732/5830 (64.0%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3738/5830 (64.1%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3743/5830 (64.2%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3749/5830 (64.3%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3755/5830 (64.4%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3761/5830 (64.5%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3767/5830 (64.6%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3773/5830 (64.7%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3778/5830 (64.8%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3784/5830 (64.9%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3790/5830 (65.0%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3796/5830 (65.1%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3802/5830 (65.2%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3807/5830 (65.3%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3813/5830 (65.4%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3819/5830 (65.5%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3825/5830 (65.6%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3831/5830 (65.7%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3837/5830 (65.8%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3842/5830 (65.9%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3848/5830 (66.0%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3854/5830 (66.1%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3860/5830 (66.2%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3866/5830 (66.3%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3872/5830 (66.4%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3877/5830 (66.5%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3883/5830 (66.6%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3889/5830 (66.7%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3895/5830 (66.8%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3901/5830 (66.9%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3907/5830 (67.0%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3912/5830 (67.1%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3918/5830 (67.2%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3924/5830 (67.3%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3930/5830 (67.4%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3936/5830 (67.5%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3942/5830 (67.6%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3947/5830 (67.7%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3953/5830 (67.8%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3959/5830 (67.9%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3965/5830 (68.0%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3971/5830 (68.1%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3977/5830 (68.2%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3982/5830 (68.3%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3988/5830 (68.4%)\n",
      "[Tue Nov 21 12:31:45 2023] Enqueueing pair 3994/5830 (68.5%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4000/5830 (68.6%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4006/5830 (68.7%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4012/5830 (68.8%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4017/5830 (68.9%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4023/5830 (69.0%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4029/5830 (69.1%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4035/5830 (69.2%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4041/5830 (69.3%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4047/5830 (69.4%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4052/5830 (69.5%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4058/5830 (69.6%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4064/5830 (69.7%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4070/5830 (69.8%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4076/5830 (69.9%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4081/5830 (70.0%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4087/5830 (70.1%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4093/5830 (70.2%)\n",
      "[Tue Nov 21 12:31:46 2023] Enqueueing pair 4099/5830 (70.3%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4105/5830 (70.4%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4111/5830 (70.5%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4116/5830 (70.6%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4122/5830 (70.7%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4128/5830 (70.8%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4134/5830 (70.9%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4140/5830 (71.0%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4146/5830 (71.1%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4151/5830 (71.2%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4157/5830 (71.3%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4163/5830 (71.4%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4169/5830 (71.5%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4175/5830 (71.6%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4181/5830 (71.7%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4186/5830 (71.8%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4192/5830 (71.9%)\n",
      "[Tue Nov 21 12:31:47 2023] Enqueueing pair 4198/5830 (72.0%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4204/5830 (72.1%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4210/5830 (72.2%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4216/5830 (72.3%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4221/5830 (72.4%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4227/5830 (72.5%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4233/5830 (72.6%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4239/5830 (72.7%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4245/5830 (72.8%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4251/5830 (72.9%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4256/5830 (73.0%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4262/5830 (73.1%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4268/5830 (73.2%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4274/5830 (73.3%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4280/5830 (73.4%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4286/5830 (73.5%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4291/5830 (73.6%)\n",
      "[Tue Nov 21 12:31:48 2023] Enqueueing pair 4297/5830 (73.7%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4303/5830 (73.8%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4309/5830 (73.9%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4315/5830 (74.0%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4321/5830 (74.1%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4326/5830 (74.2%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4332/5830 (74.3%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4338/5830 (74.4%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4344/5830 (74.5%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4350/5830 (74.6%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4356/5830 (74.7%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4361/5830 (74.8%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4367/5830 (74.9%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4373/5830 (75.0%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4379/5830 (75.1%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4385/5830 (75.2%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4390/5830 (75.3%)\n",
      "[Tue Nov 21 12:31:50 2023] Enqueueing pair 4396/5830 (75.4%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4402/5830 (75.5%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4408/5830 (75.6%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4414/5830 (75.7%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4420/5830 (75.8%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4425/5830 (75.9%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4431/5830 (76.0%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4437/5830 (76.1%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4443/5830 (76.2%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4449/5830 (76.3%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4455/5830 (76.4%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4460/5830 (76.5%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4466/5830 (76.6%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4472/5830 (76.7%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4478/5830 (76.8%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4484/5830 (76.9%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4490/5830 (77.0%)\n",
      "[Tue Nov 21 12:31:51 2023] Enqueueing pair 4495/5830 (77.1%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4501/5830 (77.2%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4507/5830 (77.3%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4513/5830 (77.4%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4519/5830 (77.5%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4525/5830 (77.6%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4530/5830 (77.7%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4536/5830 (77.8%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4542/5830 (77.9%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4548/5830 (78.0%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4554/5830 (78.1%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4560/5830 (78.2%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4565/5830 (78.3%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4571/5830 (78.4%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4577/5830 (78.5%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4583/5830 (78.6%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4589/5830 (78.7%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4595/5830 (78.8%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4600/5830 (78.9%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4606/5830 (79.0%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4612/5830 (79.1%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4618/5830 (79.2%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4624/5830 (79.3%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4630/5830 (79.4%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4635/5830 (79.5%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4641/5830 (79.6%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4647/5830 (79.7%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4653/5830 (79.8%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4659/5830 (79.9%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4664/5830 (80.0%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4670/5830 (80.1%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4676/5830 (80.2%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4682/5830 (80.3%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4688/5830 (80.4%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4694/5830 (80.5%)\n",
      "[Tue Nov 21 12:31:52 2023] Enqueueing pair 4699/5830 (80.6%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4705/5830 (80.7%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4711/5830 (80.8%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4717/5830 (80.9%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4723/5830 (81.0%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4729/5830 (81.1%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4734/5830 (81.2%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4740/5830 (81.3%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4746/5830 (81.4%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4752/5830 (81.5%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4758/5830 (81.6%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4764/5830 (81.7%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4769/5830 (81.8%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4775/5830 (81.9%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4781/5830 (82.0%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4787/5830 (82.1%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4793/5830 (82.2%)\n",
      "[Tue Nov 21 12:31:53 2023] Enqueueing pair 4799/5830 (82.3%)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4804/5830 (82.4%). Est. time left: 00:00:08 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4810/5830 (82.5%). Est. time left: 00:00:08 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4816/5830 (82.6%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4822/5830 (82.7%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4828/5830 (82.8%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4834/5830 (82.9%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4839/5830 (83.0%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4845/5830 (83.1%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4851/5830 (83.2%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4857/5830 (83.3%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4863/5830 (83.4%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4869/5830 (83.5%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4874/5830 (83.6%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4880/5830 (83.7%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4886/5830 (83.8%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4892/5830 (83.9%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:54 2023] Enqueueing pair 4898/5830 (84.0%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4904/5830 (84.1%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4909/5830 (84.2%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4915/5830 (84.3%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4921/5830 (84.4%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4927/5830 (84.5%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4933/5830 (84.6%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4939/5830 (84.7%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4944/5830 (84.8%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4950/5830 (84.9%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4956/5830 (85.0%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4962/5830 (85.1%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4968/5830 (85.2%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4973/5830 (85.3%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4979/5830 (85.4%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4985/5830 (85.5%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4991/5830 (85.6%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:56 2023] Enqueueing pair 4997/5830 (85.7%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5003/5830 (85.8%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5008/5830 (85.9%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5014/5830 (86.0%). Est. time left: 00:00:07 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5020/5830 (86.1%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5026/5830 (86.2%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5032/5830 (86.3%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5038/5830 (86.4%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5043/5830 (86.5%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5049/5830 (86.6%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5055/5830 (86.7%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5061/5830 (86.8%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5067/5830 (86.9%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5073/5830 (87.0%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5078/5830 (87.1%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5084/5830 (87.2%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5090/5830 (87.3%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5096/5830 (87.4%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5102/5830 (87.5%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5108/5830 (87.6%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5113/5830 (87.7%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5119/5830 (87.8%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5125/5830 (87.9%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5131/5830 (88.0%). Est. time left: 00:00:06 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5137/5830 (88.1%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5143/5830 (88.2%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5148/5830 (88.3%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5154/5830 (88.4%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5160/5830 (88.5%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5166/5830 (88.6%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5172/5830 (88.7%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5178/5830 (88.8%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5183/5830 (88.9%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5189/5830 (89.0%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5195/5830 (89.1%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5201/5830 (89.2%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5207/5830 (89.3%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5213/5830 (89.4%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5218/5830 (89.5%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5224/5830 (89.6%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5230/5830 (89.7%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5236/5830 (89.8%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5242/5830 (89.9%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5247/5830 (90.0%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5253/5830 (90.1%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5259/5830 (90.2%). Est. time left: 00:00:05 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5265/5830 (90.3%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5271/5830 (90.4%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5277/5830 (90.5%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5282/5830 (90.6%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5288/5830 (90.7%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:58 2023] Enqueueing pair 5294/5830 (90.8%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5300/5830 (90.9%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5306/5830 (91.0%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5312/5830 (91.1%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5317/5830 (91.2%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5323/5830 (91.3%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5329/5830 (91.4%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5335/5830 (91.5%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5341/5830 (91.6%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5347/5830 (91.7%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5352/5830 (91.8%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5358/5830 (91.9%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5364/5830 (92.0%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5370/5830 (92.1%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5376/5830 (92.2%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5382/5830 (92.3%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5387/5830 (92.4%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5393/5830 (92.5%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5399/5830 (92.6%). Est. time left: 00:00:04 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5405/5830 (92.7%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5411/5830 (92.8%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5417/5830 (92.9%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5422/5830 (93.0%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5428/5830 (93.1%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5434/5830 (93.2%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5440/5830 (93.3%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5446/5830 (93.4%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5452/5830 (93.5%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5457/5830 (93.6%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5463/5830 (93.7%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5469/5830 (93.8%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5475/5830 (93.9%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5481/5830 (94.0%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5487/5830 (94.1%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5492/5830 (94.2%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:31:59 2023] Enqueueing pair 5498/5830 (94.3%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5504/5830 (94.4%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5510/5830 (94.5%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5516/5830 (94.6%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5522/5830 (94.7%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5527/5830 (94.8%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5533/5830 (94.9%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5539/5830 (95.0%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5545/5830 (95.1%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5551/5830 (95.2%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5556/5830 (95.3%). Est. time left: 00:00:03 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5562/5830 (95.4%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5568/5830 (95.5%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5574/5830 (95.6%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5580/5830 (95.7%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5586/5830 (95.8%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5591/5830 (95.9%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:02 2023] Enqueueing pair 5597/5830 (96.0%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5603/5830 (96.1%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5609/5830 (96.2%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5615/5830 (96.3%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5621/5830 (96.4%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5626/5830 (96.5%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5632/5830 (96.6%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5638/5830 (96.7%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5644/5830 (96.8%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5650/5830 (96.9%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5656/5830 (97.0%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5661/5830 (97.1%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5667/5830 (97.2%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5673/5830 (97.3%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5679/5830 (97.4%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5685/5830 (97.5%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5691/5830 (97.6%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5696/5830 (97.7%). Est. time left: 00:00:02 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5702/5830 (97.8%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5708/5830 (97.9%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5714/5830 (98.0%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5720/5830 (98.1%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5726/5830 (98.2%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5731/5830 (98.3%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5737/5830 (98.4%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5743/5830 (98.5%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5749/5830 (98.6%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5755/5830 (98.7%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5761/5830 (98.8%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5766/5830 (98.9%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5772/5830 (99.0%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5778/5830 (99.1%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5784/5830 (99.2%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5790/5830 (99.3%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5796/5830 (99.4%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5801/5830 (99.5%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5807/5830 (99.6%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5813/5830 (99.7%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5819/5830 (99.8%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:04 2023] Enqueueing pair 5825/5830 (99.9%). Est. time left: 00:00:01 (hh:mm:ss)\n",
      "[Tue Nov 21 12:32:23 2023] 5830/5830 (100%) pairs aligned\n",
      "[Tue Nov 21 12:32:23 2023] Alignment of pairs finished\n",
      "[Tue Nov 21 12:32:23 2023] Translating coordinates from local pair level to contig level\n",
      "[Tue Nov 21 12:32:24 2023] Finished spliced alignment\n",
      "[Tue Nov 21 12:32:24 2023] Flagging top chains\n",
      "[Tue Nov 21 12:32:24 2023] Processing the output\n",
      "[Tue Nov 21 12:32:25 2023] Output processed\n",
      "[Tue Nov 21 12:32:25 2023] ProtHint finished.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#*********\n",
      "# WARNING: Number of reliable training genes is low (200). Recommended are at least 600 genes\n",
      "#*********\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ProtHint Version 2.6.0\n",
      "Copyright 2019, Georgia Institute of Technology, USA\n",
      "\n",
      "Please cite\n",
      "  - ProtHint: https://doi.org/10.1093/nargab/lqaa026\n",
      "  - DIAMOND:  https://doi.org/10.1038/nmeth.3176\n",
      "  - Spaln:    https://doi.org/10.1093/bioinformatics/btn460\n",
      "\n",
      "Called from: /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2\n",
      "Cmd: /opt/ETP/bin/gmes/ProtHint/bin/prothint.py --threads=6 --geneSeeds /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/augustus.hints_iter1.gtf --prevGeneSeeds /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/GeneMark-ES/genemark.gtf --prevSpalnGff /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/Spaln/spaln_iter1.gff /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/genome.fa /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/proteins.fa\n",
      "\n",
      "[Tue Nov 21 12:34:17 2023] Pre-processing protein input\n",
      "ProtHint is running in the iterative mode.\n",
      "[Tue Nov 21 12:34:17 2023] Selecting a subset of data to run in the next iteration\n",
      "[Tue Nov 21 12:34:17 2023] Translating gene seeds to proteins\n",
      "[Tue Nov 21 12:34:18 2023] Translation of seeds finished\n",
      "[Tue Nov 21 12:34:18 2023] Running DIAMOND\n",
      "diamond v0.9.24.125 | by Benjamin Buchfink <buchfink@gmail.com>\n",
      "Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>\n",
      "Check http://github.com/bbuchfink/diamond for updates.\n",
      "\n",
      "#CPU threads: 6\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Database file: /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER2/proty_m3s4nc\n",
      "Opening the database file...  [0.000982s]\n",
      "Loading sequences...  [0.017125s]\n",
      "Masking sequences...  [0.125809s]\n",
      "Writing sequences...  [0.00624s]\n",
      "Hashing sequences...  [0.002115s]\n",
      "Loading sequences...  [2.1e-05s]\n",
      "Writing trailer...  [0.000179s]\n",
      "Closing the input file...  [2.4e-05s]\n",
      "Closing the database file...  [0.001597s]\n",
      "Database hash = e1048f69f134493ac567793c01fff94c\n",
      "Processed 5733 sequences, 2599346 letters.\n",
      "Total time = 0.154299s\n",
      "diamond v0.9.24.125 | by Benjamin Buchfink <buchfink@gmail.com>\n",
      "Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>\n",
      "Check http://github.com/bbuchfink/diamond for updates.\n",
      "\n",
      "#CPU threads: 6\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: \n",
      "Opening the database...  [3.1e-05s]\n",
      "#Target sequences to report alignments for: 25\n",
      "Opening the input file...  [1.5e-05s]\n",
      "Opening the output file...  [4.7e-05s]\n",
      "Loading query sequences...  [0.000368s]\n",
      "Masking queries...  [0.008409s]\n",
      "Building query seed set...  [0.002186s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.000962s]\n",
      "Allocating buffers...  [3e-05s]\n",
      "Loading reference sequences...  [0.006163s]\n",
      "Building reference histograms...  [0.022736s]\n",
      "Allocating buffers...  [5.5e-05s]\n",
      "Initializing temporary storage...  [0.003317s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 0.\n",
      "Building reference index...  [0.031544s]\n",
      "Building query index...  [0.002146s]\n",
      "Building seed filter...  [0.002951s]\n",
      "Searching alignments...  [0.021384s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 1.\n",
      "Building reference index...  [0.032472s]\n",
      "Building query index...  [0.0022s]\n",
      "Building seed filter...  [0.00289s]\n",
      "Searching alignments...  [0.016015s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 2.\n",
      "Building reference index...  [0.035953s]\n",
      "Building query index...  [0.002235s]\n",
      "Building seed filter...  [0.003249s]\n",
      "Searching alignments...  [0.016627s]\n",
      "Processing query chunk 0, reference chunk 0, shape 0, index chunk 3.\n",
      "Building reference index...  [0.032635s]\n",
      "Building query index...  [0.002026s]\n",
      "Building seed filter...  [0.00304s]\n",
      "Searching alignments...  [0.0164s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 0.\n",
      "Building reference index...  [0.033758s]\n",
      "Building query index...  [0.002103s]\n",
      "Building seed filter...  [0.003369s]\n",
      "Searching alignments...  [0.01306s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 1.\n",
      "Building reference index...  [0.036412s]\n",
      "Building query index...  [0.002689s]\n",
      "Building seed filter...  [0.002792s]\n",
      "Searching alignments...  [0.016124s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 2.\n",
      "Building reference index...  [0.035249s]\n",
      "Building query index...  [0.002256s]\n",
      "Building seed filter...  [0.003291s]\n",
      "Searching alignments...  [0.016078s]\n",
      "Processing query chunk 0, reference chunk 0, shape 1, index chunk 3.\n",
      "Building reference index...  [0.029207s]\n",
      "Building query index...  [0.001877s]\n",
      "Building seed filter...  [0.003316s]\n",
      "Searching alignments...  [0.015838s]\n",
      "Deallocating buffers...  [0.000594s]\n",
      "Computing alignments...  [0.116782s]\n",
      "Deallocating reference...  [0.000312s]\n",
      "Loading reference sequences...  [2e-05s]\n",
      "Deallocating buffers...  [4e-06s]\n",
      "Deallocating queries...  [3e-05s]\n",
      "Loading query sequences...  [1.8e-05s]\n",
      "Closing the input file...  [1.6e-05s]\n",
      "Closing the output file...  [0.000317s]\n",
      "Closing the database file...  [1.2e-05s]\n",
      "Deallocating taxonomy...  [4e-06s]\n",
      "Total time = 0.605254s\n",
      "Reported 3443 pairwise alignments, 3443 HSPs.\n",
      "144 queries aligned.\n",
      "[Tue Nov 21 12:34:18 2023] DIAMOND finished\n",
      "[Tue Nov 21 12:34:18 2023] Preparing pairs for alignments\n",
      "[Tue Nov 21 12:34:19 2023] Preparation of pairs finished\n",
      "[Tue Nov 21 12:34:19 2023] Starting spliced alignment with spaln\n",
      "[Tue Nov 21 12:34:19 2023] Loading alignment pairs into memory\n",
      "[Tue Nov 21 12:34:19 2023] Pairs loaded. Number of pairs to align: 3443\n",
      "[Tue Nov 21 12:34:19 2023] Starting the alignments\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 0/3443 (0.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 4/3443 (0.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 7/3443 (0.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 11/3443 (0.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 14/3443 (0.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 18/3443 (0.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 21/3443 (0.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 25/3443 (0.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 28/3443 (0.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 31/3443 (0.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 35/3443 (1.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 38/3443 (1.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 42/3443 (1.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 45/3443 (1.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 49/3443 (1.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 52/3443 (1.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 56/3443 (1.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 59/3443 (1.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 62/3443 (1.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 66/3443 (1.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 69/3443 (2.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 73/3443 (2.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 76/3443 (2.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 80/3443 (2.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 83/3443 (2.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 87/3443 (2.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 90/3443 (2.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 93/3443 (2.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 97/3443 (2.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 100/3443 (2.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 104/3443 (3.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 107/3443 (3.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 111/3443 (3.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 114/3443 (3.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 118/3443 (3.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 121/3443 (3.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 124/3443 (3.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 128/3443 (3.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 131/3443 (3.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 135/3443 (3.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 138/3443 (4.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 142/3443 (4.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 145/3443 (4.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 149/3443 (4.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 152/3443 (4.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 155/3443 (4.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 159/3443 (4.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 162/3443 (4.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 166/3443 (4.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 169/3443 (4.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 173/3443 (5.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 176/3443 (5.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 180/3443 (5.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 183/3443 (5.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 186/3443 (5.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 190/3443 (5.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 193/3443 (5.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 197/3443 (5.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 200/3443 (5.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 204/3443 (5.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 207/3443 (6.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 211/3443 (6.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 214/3443 (6.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 217/3443 (6.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 221/3443 (6.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 224/3443 (6.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 228/3443 (6.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 231/3443 (6.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 235/3443 (6.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 238/3443 (6.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 242/3443 (7.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 245/3443 (7.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 248/3443 (7.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 252/3443 (7.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 255/3443 (7.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 259/3443 (7.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 262/3443 (7.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 266/3443 (7.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 269/3443 (7.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 272/3443 (7.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 276/3443 (8.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 279/3443 (8.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 283/3443 (8.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 286/3443 (8.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 290/3443 (8.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 293/3443 (8.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 297/3443 (8.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 300/3443 (8.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 303/3443 (8.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 307/3443 (8.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 310/3443 (9.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 314/3443 (9.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 317/3443 (9.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 321/3443 (9.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 324/3443 (9.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 328/3443 (9.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 331/3443 (9.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 334/3443 (9.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 338/3443 (9.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 341/3443 (9.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 345/3443 (10.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 348/3443 (10.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 352/3443 (10.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 355/3443 (10.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 359/3443 (10.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 362/3443 (10.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 365/3443 (10.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 369/3443 (10.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 372/3443 (10.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 376/3443 (10.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 379/3443 (11.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 383/3443 (11.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 386/3443 (11.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 390/3443 (11.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 393/3443 (11.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 396/3443 (11.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 400/3443 (11.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 403/3443 (11.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 407/3443 (11.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 410/3443 (11.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 414/3443 (12.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 417/3443 (12.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 421/3443 (12.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 424/3443 (12.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 427/3443 (12.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 431/3443 (12.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 434/3443 (12.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 438/3443 (12.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 441/3443 (12.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 445/3443 (12.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 448/3443 (13.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 452/3443 (13.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 455/3443 (13.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 458/3443 (13.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 462/3443 (13.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 465/3443 (13.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 469/3443 (13.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 472/3443 (13.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 476/3443 (13.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 479/3443 (13.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 483/3443 (14.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 486/3443 (14.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 489/3443 (14.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 493/3443 (14.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 496/3443 (14.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 500/3443 (14.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 503/3443 (14.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 507/3443 (14.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 510/3443 (14.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 514/3443 (14.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 517/3443 (15.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 520/3443 (15.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 524/3443 (15.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 527/3443 (15.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 531/3443 (15.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 534/3443 (15.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 538/3443 (15.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 541/3443 (15.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 544/3443 (15.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 548/3443 (15.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 551/3443 (16.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 555/3443 (16.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 558/3443 (16.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 562/3443 (16.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 565/3443 (16.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 569/3443 (16.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 572/3443 (16.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 575/3443 (16.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 579/3443 (16.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 582/3443 (16.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 586/3443 (17.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 589/3443 (17.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 593/3443 (17.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 596/3443 (17.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 600/3443 (17.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 603/3443 (17.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 606/3443 (17.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 610/3443 (17.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 613/3443 (17.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 617/3443 (17.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 620/3443 (18.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 624/3443 (18.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 627/3443 (18.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 631/3443 (18.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 634/3443 (18.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 637/3443 (18.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 641/3443 (18.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 644/3443 (18.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 648/3443 (18.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 651/3443 (18.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 655/3443 (19.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 658/3443 (19.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 662/3443 (19.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 665/3443 (19.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 668/3443 (19.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 672/3443 (19.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 675/3443 (19.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 679/3443 (19.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 682/3443 (19.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 686/3443 (19.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 689/3443 (20.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 693/3443 (20.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 696/3443 (20.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 699/3443 (20.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 703/3443 (20.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 706/3443 (20.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 710/3443 (20.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 713/3443 (20.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 717/3443 (20.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 720/3443 (20.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 724/3443 (21.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 727/3443 (21.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 730/3443 (21.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 734/3443 (21.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 737/3443 (21.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 741/3443 (21.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 744/3443 (21.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 748/3443 (21.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 751/3443 (21.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 755/3443 (21.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 758/3443 (22.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 761/3443 (22.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 765/3443 (22.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 768/3443 (22.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 772/3443 (22.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 775/3443 (22.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 779/3443 (22.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 782/3443 (22.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 786/3443 (22.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 789/3443 (22.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 792/3443 (23.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 796/3443 (23.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 799/3443 (23.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 803/3443 (23.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 806/3443 (23.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 810/3443 (23.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 813/3443 (23.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 816/3443 (23.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 820/3443 (23.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 823/3443 (23.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 827/3443 (24.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 830/3443 (24.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 834/3443 (24.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 837/3443 (24.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 841/3443 (24.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 844/3443 (24.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 847/3443 (24.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 851/3443 (24.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 854/3443 (24.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 858/3443 (24.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 861/3443 (25.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 865/3443 (25.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 868/3443 (25.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 872/3443 (25.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 875/3443 (25.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 878/3443 (25.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 882/3443 (25.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 885/3443 (25.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 889/3443 (25.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 892/3443 (25.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 896/3443 (26.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 899/3443 (26.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 903/3443 (26.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 906/3443 (26.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 909/3443 (26.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 913/3443 (26.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 916/3443 (26.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 920/3443 (26.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 923/3443 (26.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 927/3443 (26.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 930/3443 (27.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 934/3443 (27.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 937/3443 (27.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 940/3443 (27.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 944/3443 (27.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 947/3443 (27.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 951/3443 (27.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 954/3443 (27.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 958/3443 (27.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 961/3443 (27.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 965/3443 (28.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 968/3443 (28.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 971/3443 (28.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 975/3443 (28.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 978/3443 (28.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 982/3443 (28.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 985/3443 (28.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 989/3443 (28.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 992/3443 (28.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 996/3443 (28.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 999/3443 (29.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1002/3443 (29.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1006/3443 (29.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1009/3443 (29.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1013/3443 (29.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1016/3443 (29.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1020/3443 (29.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1023/3443 (29.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1027/3443 (29.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1030/3443 (29.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1033/3443 (30.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1037/3443 (30.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1040/3443 (30.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1044/3443 (30.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1047/3443 (30.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1051/3443 (30.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1054/3443 (30.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1058/3443 (30.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1061/3443 (30.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1064/3443 (30.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1068/3443 (31.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1071/3443 (31.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1075/3443 (31.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1078/3443 (31.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1082/3443 (31.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1085/3443 (31.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1088/3443 (31.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1092/3443 (31.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1095/3443 (31.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1099/3443 (31.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1102/3443 (32.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1106/3443 (32.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1109/3443 (32.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1113/3443 (32.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1116/3443 (32.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1119/3443 (32.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1123/3443 (32.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1126/3443 (32.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1130/3443 (32.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1133/3443 (32.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1137/3443 (33.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1140/3443 (33.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1144/3443 (33.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1147/3443 (33.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1150/3443 (33.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1154/3443 (33.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1157/3443 (33.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1161/3443 (33.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1164/3443 (33.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1168/3443 (33.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1171/3443 (34.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1175/3443 (34.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1178/3443 (34.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1181/3443 (34.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1185/3443 (34.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1188/3443 (34.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1192/3443 (34.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1195/3443 (34.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1199/3443 (34.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1202/3443 (34.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1206/3443 (35.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1209/3443 (35.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1212/3443 (35.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1216/3443 (35.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1219/3443 (35.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1223/3443 (35.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1226/3443 (35.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1230/3443 (35.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1233/3443 (35.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1237/3443 (35.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1240/3443 (36.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1243/3443 (36.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1247/3443 (36.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1250/3443 (36.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1254/3443 (36.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1257/3443 (36.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1261/3443 (36.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1264/3443 (36.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1268/3443 (36.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1271/3443 (36.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1274/3443 (37.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1278/3443 (37.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1281/3443 (37.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1285/3443 (37.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1288/3443 (37.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1292/3443 (37.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1295/3443 (37.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1299/3443 (37.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1302/3443 (37.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1305/3443 (37.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1309/3443 (38.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1312/3443 (38.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1316/3443 (38.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1319/3443 (38.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1323/3443 (38.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1326/3443 (38.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1329/3443 (38.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1333/3443 (38.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1336/3443 (38.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1340/3443 (38.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1343/3443 (39.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1347/3443 (39.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1350/3443 (39.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1354/3443 (39.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1357/3443 (39.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1360/3443 (39.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1364/3443 (39.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1367/3443 (39.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1371/3443 (39.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1374/3443 (39.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1378/3443 (40.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1381/3443 (40.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1385/3443 (40.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1388/3443 (40.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1391/3443 (40.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1395/3443 (40.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1398/3443 (40.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1402/3443 (40.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1405/3443 (40.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1409/3443 (40.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1412/3443 (41.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1416/3443 (41.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1419/3443 (41.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1422/3443 (41.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1426/3443 (41.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1429/3443 (41.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1433/3443 (41.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1436/3443 (41.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1440/3443 (41.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1443/3443 (41.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1447/3443 (42.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1450/3443 (42.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1453/3443 (42.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1457/3443 (42.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1460/3443 (42.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1464/3443 (42.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1467/3443 (42.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1471/3443 (42.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1474/3443 (42.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1478/3443 (42.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1481/3443 (43.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1484/3443 (43.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1488/3443 (43.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1491/3443 (43.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1495/3443 (43.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1498/3443 (43.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1502/3443 (43.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1505/3443 (43.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1509/3443 (43.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1512/3443 (43.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1515/3443 (44.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1519/3443 (44.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1522/3443 (44.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1526/3443 (44.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1529/3443 (44.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1533/3443 (44.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1536/3443 (44.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1540/3443 (44.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1543/3443 (44.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1546/3443 (44.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1550/3443 (45.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1553/3443 (45.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1557/3443 (45.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1560/3443 (45.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1564/3443 (45.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1567/3443 (45.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1571/3443 (45.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1574/3443 (45.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1577/3443 (45.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1581/3443 (45.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1584/3443 (46.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1588/3443 (46.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1591/3443 (46.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1595/3443 (46.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1598/3443 (46.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1601/3443 (46.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1605/3443 (46.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1608/3443 (46.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1612/3443 (46.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1615/3443 (46.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1619/3443 (47.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1622/3443 (47.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1626/3443 (47.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1629/3443 (47.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1632/3443 (47.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1636/3443 (47.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1639/3443 (47.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1643/3443 (47.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1646/3443 (47.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1650/3443 (47.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1653/3443 (48.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1657/3443 (48.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1660/3443 (48.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1663/3443 (48.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1667/3443 (48.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1670/3443 (48.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1674/3443 (48.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1677/3443 (48.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1681/3443 (48.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1684/3443 (48.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1688/3443 (49.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1691/3443 (49.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1694/3443 (49.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1698/3443 (49.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1701/3443 (49.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1705/3443 (49.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1708/3443 (49.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1712/3443 (49.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1715/3443 (49.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1719/3443 (49.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1722/3443 (50.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1725/3443 (50.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1729/3443 (50.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1732/3443 (50.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1736/3443 (50.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1739/3443 (50.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1743/3443 (50.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1746/3443 (50.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1750/3443 (50.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1753/3443 (50.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1756/3443 (51.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1760/3443 (51.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1763/3443 (51.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1767/3443 (51.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1770/3443 (51.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1774/3443 (51.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1777/3443 (51.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1781/3443 (51.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1784/3443 (51.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1787/3443 (51.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1791/3443 (52.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1794/3443 (52.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1798/3443 (52.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1801/3443 (52.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1805/3443 (52.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1808/3443 (52.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1812/3443 (52.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1815/3443 (52.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1818/3443 (52.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1822/3443 (52.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1825/3443 (53.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1829/3443 (53.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1832/3443 (53.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1836/3443 (53.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1839/3443 (53.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1843/3443 (53.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1846/3443 (53.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1849/3443 (53.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1853/3443 (53.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1856/3443 (53.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1860/3443 (54.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1863/3443 (54.1%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1867/3443 (54.2%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1870/3443 (54.3%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1873/3443 (54.4%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1877/3443 (54.5%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1880/3443 (54.6%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1884/3443 (54.7%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1887/3443 (54.8%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1891/3443 (54.9%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1894/3443 (55.0%)\n",
      "[Tue Nov 21 12:34:19 2023] Enqueueing pair 1898/3443 (55.1%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1901/3443 (55.2%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1904/3443 (55.3%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1908/3443 (55.4%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1911/3443 (55.5%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1915/3443 (55.6%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1918/3443 (55.7%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1922/3443 (55.8%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1925/3443 (55.9%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1929/3443 (56.0%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1932/3443 (56.1%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1935/3443 (56.2%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1939/3443 (56.3%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1942/3443 (56.4%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1946/3443 (56.5%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1949/3443 (56.6%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1953/3443 (56.7%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1956/3443 (56.8%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1960/3443 (56.9%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1963/3443 (57.0%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1966/3443 (57.1%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1970/3443 (57.2%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1973/3443 (57.3%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1977/3443 (57.4%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1980/3443 (57.5%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1984/3443 (57.6%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1987/3443 (57.7%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1991/3443 (57.8%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1994/3443 (57.9%)\n",
      "[Tue Nov 21 12:34:25 2023] Enqueueing pair 1997/3443 (58.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2001/3443 (58.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2004/3443 (58.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2008/3443 (58.3%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2011/3443 (58.4%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2015/3443 (58.5%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2018/3443 (58.6%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2022/3443 (58.7%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2025/3443 (58.8%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2028/3443 (58.9%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2032/3443 (59.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2035/3443 (59.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2039/3443 (59.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2042/3443 (59.3%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2046/3443 (59.4%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2049/3443 (59.5%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2053/3443 (59.6%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2056/3443 (59.7%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2059/3443 (59.8%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2063/3443 (59.9%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2066/3443 (60.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2070/3443 (60.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2073/3443 (60.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2077/3443 (60.3%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2080/3443 (60.4%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2084/3443 (60.5%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2087/3443 (60.6%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2090/3443 (60.7%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2094/3443 (60.8%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2097/3443 (60.9%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2101/3443 (61.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2104/3443 (61.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2108/3443 (61.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2111/3443 (61.3%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2115/3443 (61.4%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2118/3443 (61.5%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2121/3443 (61.6%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2125/3443 (61.7%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2128/3443 (61.8%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2132/3443 (61.9%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2135/3443 (62.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2139/3443 (62.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2142/3443 (62.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2145/3443 (62.3%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2149/3443 (62.4%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2152/3443 (62.5%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2156/3443 (62.6%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2159/3443 (62.7%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2163/3443 (62.8%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2166/3443 (62.9%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2170/3443 (63.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2173/3443 (63.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2176/3443 (63.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2180/3443 (63.3%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2183/3443 (63.4%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2187/3443 (63.5%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2190/3443 (63.6%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2194/3443 (63.7%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2197/3443 (63.8%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2201/3443 (63.9%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2204/3443 (64.0%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2207/3443 (64.1%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2211/3443 (64.2%)\n",
      "[Tue Nov 21 12:34:26 2023] Enqueueing pair 2214/3443 (64.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2218/3443 (64.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2221/3443 (64.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2225/3443 (64.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2228/3443 (64.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2232/3443 (64.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2235/3443 (64.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2238/3443 (65.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2242/3443 (65.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2245/3443 (65.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2249/3443 (65.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2252/3443 (65.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2256/3443 (65.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2259/3443 (65.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2263/3443 (65.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2266/3443 (65.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2269/3443 (65.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2273/3443 (66.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2276/3443 (66.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2280/3443 (66.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2283/3443 (66.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2287/3443 (66.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2290/3443 (66.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2294/3443 (66.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2297/3443 (66.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2300/3443 (66.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2304/3443 (66.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2307/3443 (67.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2311/3443 (67.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2314/3443 (67.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2318/3443 (67.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2321/3443 (67.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2325/3443 (67.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2328/3443 (67.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2331/3443 (67.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2335/3443 (67.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2338/3443 (67.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2342/3443 (68.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2345/3443 (68.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2349/3443 (68.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2352/3443 (68.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2356/3443 (68.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2359/3443 (68.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2362/3443 (68.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2366/3443 (68.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2369/3443 (68.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2373/3443 (68.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2376/3443 (69.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2380/3443 (69.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2383/3443 (69.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2386/3443 (69.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2390/3443 (69.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2393/3443 (69.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2397/3443 (69.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2400/3443 (69.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2404/3443 (69.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2407/3443 (69.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2411/3443 (70.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2414/3443 (70.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2417/3443 (70.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2421/3443 (70.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2424/3443 (70.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2428/3443 (70.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2431/3443 (70.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2435/3443 (70.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2438/3443 (70.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2442/3443 (70.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2445/3443 (71.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2448/3443 (71.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2452/3443 (71.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2455/3443 (71.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2459/3443 (71.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2462/3443 (71.5%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2466/3443 (71.6%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2469/3443 (71.7%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2473/3443 (71.8%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2476/3443 (71.9%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2479/3443 (72.0%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2483/3443 (72.1%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2486/3443 (72.2%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2490/3443 (72.3%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2493/3443 (72.4%)\n",
      "[Tue Nov 21 12:34:27 2023] Enqueueing pair 2497/3443 (72.5%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2500/3443 (72.6%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2504/3443 (72.7%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2507/3443 (72.8%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2510/3443 (72.9%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2514/3443 (73.0%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2517/3443 (73.1%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2521/3443 (73.2%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2524/3443 (73.3%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2528/3443 (73.4%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2531/3443 (73.5%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2535/3443 (73.6%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2538/3443 (73.7%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2541/3443 (73.8%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2545/3443 (73.9%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2548/3443 (74.0%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2552/3443 (74.1%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2555/3443 (74.2%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2559/3443 (74.3%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2562/3443 (74.4%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2566/3443 (74.5%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2569/3443 (74.6%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2572/3443 (74.7%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2576/3443 (74.8%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2579/3443 (74.9%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2583/3443 (75.0%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2586/3443 (75.1%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2590/3443 (75.2%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2593/3443 (75.3%)\n",
      "[Tue Nov 21 12:34:31 2023] Enqueueing pair 2597/3443 (75.4%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2600/3443 (75.5%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2603/3443 (75.6%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2607/3443 (75.7%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2610/3443 (75.8%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2614/3443 (75.9%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2617/3443 (76.0%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2621/3443 (76.1%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2624/3443 (76.2%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2628/3443 (76.3%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2631/3443 (76.4%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2634/3443 (76.5%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2638/3443 (76.6%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2641/3443 (76.7%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2645/3443 (76.8%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2648/3443 (76.9%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2652/3443 (77.0%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2655/3443 (77.1%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2658/3443 (77.2%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2662/3443 (77.3%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2665/3443 (77.4%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2669/3443 (77.5%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2672/3443 (77.6%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2676/3443 (77.7%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2679/3443 (77.8%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2683/3443 (77.9%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2686/3443 (78.0%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2689/3443 (78.1%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2693/3443 (78.2%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2696/3443 (78.3%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2700/3443 (78.4%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2703/3443 (78.5%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2707/3443 (78.6%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2710/3443 (78.7%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2714/3443 (78.8%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2717/3443 (78.9%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2720/3443 (79.0%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2724/3443 (79.1%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2727/3443 (79.2%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2731/3443 (79.3%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2734/3443 (79.4%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2738/3443 (79.5%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2741/3443 (79.6%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2745/3443 (79.7%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2748/3443 (79.8%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2751/3443 (79.9%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2755/3443 (80.0%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2758/3443 (80.1%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2762/3443 (80.2%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2765/3443 (80.3%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2769/3443 (80.4%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2772/3443 (80.5%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2776/3443 (80.6%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2779/3443 (80.7%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2782/3443 (80.8%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2786/3443 (80.9%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2789/3443 (81.0%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2793/3443 (81.1%)\n",
      "[Tue Nov 21 12:34:32 2023] Enqueueing pair 2796/3443 (81.2%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2800/3443 (81.3%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2803/3443 (81.4%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2807/3443 (81.5%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2810/3443 (81.6%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2813/3443 (81.7%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2817/3443 (81.8%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2820/3443 (81.9%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2824/3443 (82.0%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2827/3443 (82.1%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2831/3443 (82.2%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2834/3443 (82.3%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2838/3443 (82.4%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2841/3443 (82.5%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2844/3443 (82.6%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2848/3443 (82.7%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2851/3443 (82.8%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2855/3443 (82.9%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2858/3443 (83.0%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2862/3443 (83.1%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2865/3443 (83.2%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2869/3443 (83.3%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2872/3443 (83.4%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2875/3443 (83.5%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2879/3443 (83.6%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2882/3443 (83.7%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2886/3443 (83.8%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2889/3443 (83.9%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2893/3443 (84.0%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2896/3443 (84.1%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2900/3443 (84.2%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2903/3443 (84.3%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2906/3443 (84.4%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2910/3443 (84.5%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2913/3443 (84.6%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2917/3443 (84.7%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2920/3443 (84.8%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2924/3443 (84.9%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2927/3443 (85.0%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2930/3443 (85.1%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2934/3443 (85.2%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2937/3443 (85.3%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2941/3443 (85.4%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2944/3443 (85.5%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2948/3443 (85.6%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2951/3443 (85.7%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2955/3443 (85.8%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2958/3443 (85.9%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2961/3443 (86.0%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2965/3443 (86.1%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2968/3443 (86.2%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2972/3443 (86.3%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2975/3443 (86.4%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2979/3443 (86.5%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2982/3443 (86.6%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2986/3443 (86.7%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2989/3443 (86.8%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2992/3443 (86.9%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2996/3443 (87.0%)\n",
      "[Tue Nov 21 12:34:33 2023] Enqueueing pair 2999/3443 (87.1%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3003/3443 (87.2%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3006/3443 (87.3%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3010/3443 (87.4%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3013/3443 (87.5%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3017/3443 (87.6%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3020/3443 (87.7%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3023/3443 (87.8%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3027/3443 (87.9%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3030/3443 (88.0%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3034/3443 (88.1%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3037/3443 (88.2%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3041/3443 (88.3%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3044/3443 (88.4%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3048/3443 (88.5%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3051/3443 (88.6%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3054/3443 (88.7%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3058/3443 (88.8%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3061/3443 (88.9%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3065/3443 (89.0%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3068/3443 (89.1%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3072/3443 (89.2%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3075/3443 (89.3%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3079/3443 (89.4%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3082/3443 (89.5%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3085/3443 (89.6%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3089/3443 (89.7%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3092/3443 (89.8%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3096/3443 (89.9%)\n",
      "[Tue Nov 21 12:34:34 2023] Enqueueing pair 3099/3443 (90.0%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3103/3443 (90.1%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3106/3443 (90.2%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3110/3443 (90.3%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3113/3443 (90.4%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3116/3443 (90.5%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3120/3443 (90.6%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3123/3443 (90.7%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3127/3443 (90.8%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3130/3443 (90.9%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3134/3443 (91.0%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3137/3443 (91.1%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3141/3443 (91.2%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3144/3443 (91.3%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3147/3443 (91.4%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3151/3443 (91.5%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3154/3443 (91.6%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3158/3443 (91.7%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3161/3443 (91.8%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3165/3443 (91.9%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3168/3443 (92.0%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3172/3443 (92.1%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3175/3443 (92.2%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3178/3443 (92.3%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3182/3443 (92.4%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3185/3443 (92.5%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3189/3443 (92.6%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3192/3443 (92.7%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3196/3443 (92.8%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3199/3443 (92.9%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3202/3443 (93.0%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3206/3443 (93.1%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3209/3443 (93.2%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3213/3443 (93.3%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3216/3443 (93.4%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3220/3443 (93.5%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3223/3443 (93.6%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3227/3443 (93.7%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3230/3443 (93.8%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3233/3443 (93.9%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3237/3443 (94.0%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3240/3443 (94.1%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3244/3443 (94.2%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3247/3443 (94.3%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3251/3443 (94.4%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3254/3443 (94.5%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3258/3443 (94.6%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3261/3443 (94.7%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3264/3443 (94.8%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3268/3443 (94.9%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3271/3443 (95.0%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3275/3443 (95.1%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3278/3443 (95.2%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3282/3443 (95.3%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3285/3443 (95.4%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3289/3443 (95.5%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3292/3443 (95.6%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3295/3443 (95.7%)\n",
      "[Tue Nov 21 12:34:37 2023] Enqueueing pair 3299/3443 (95.8%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3302/3443 (95.9%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3306/3443 (96.0%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3309/3443 (96.1%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3313/3443 (96.2%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3316/3443 (96.3%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3320/3443 (96.4%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3323/3443 (96.5%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3326/3443 (96.6%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3330/3443 (96.7%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3333/3443 (96.8%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3337/3443 (96.9%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3340/3443 (97.0%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3344/3443 (97.1%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3347/3443 (97.2%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3351/3443 (97.3%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3354/3443 (97.4%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3357/3443 (97.5%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3361/3443 (97.6%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3364/3443 (97.7%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3368/3443 (97.8%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3371/3443 (97.9%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3375/3443 (98.0%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3378/3443 (98.1%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3382/3443 (98.2%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3385/3443 (98.3%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3388/3443 (98.4%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3392/3443 (98.5%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3395/3443 (98.6%)\n",
      "[Tue Nov 21 12:34:38 2023] Enqueueing pair 3399/3443 (98.7%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3402/3443 (98.8%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3406/3443 (98.9%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3409/3443 (99.0%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3413/3443 (99.1%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3416/3443 (99.2%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3419/3443 (99.3%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3423/3443 (99.4%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3426/3443 (99.5%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3430/3443 (99.6%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3433/3443 (99.7%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3437/3443 (99.8%)\n",
      "[Tue Nov 21 12:34:39 2023] Enqueueing pair 3440/3443 (99.9%)\n",
      "[Tue Nov 21 12:34:57 2023] 3443/3443 (100%) pairs aligned\n",
      "[Tue Nov 21 12:34:57 2023] Alignment of pairs finished\n",
      "[Tue Nov 21 12:34:57 2023] Translating coordinates from local pair level to contig level\n",
      "[Tue Nov 21 12:34:57 2023] Finished spliced alignment\n",
      "[Tue Nov 21 12:34:57 2023] Flagging top chains\n",
      "[Tue Nov 21 12:34:57 2023] Processing the output\n",
      "[Tue Nov 21 12:34:58 2023] Output processed\n",
      "[Tue Nov 21 12:34:58 2023] ProtHint finished.\n",
      "\n",
      "real\t5m38.279s\n",
      "user\t16m30.903s\n",
      "sys\t1m59.146s\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "T=6 # adjust to number of threads that you booted with\n",
    "\n",
    "ORTHODB=/opt/BRAKER/example/proteins.fa # adjust to suitable OrthoDB clade, see BRAKER3\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d BRAKER2 ]\n",
    "then\n",
    "    rm -rf BRAKER2\n",
    "fi\n",
    "\n",
    "time braker.pl --workingdir=BRAKER2 --genome=/opt/BRAKER/example/genome.fa --prot_seq=${ORTHODB} \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --gm_max_intergenic 10000 --skipOptimize \\ # remember to remove both options if you are running a real job\n",
    "    2> braker2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ff71c1e",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Out of time, job died? Click here.</b></summary>\n",
    "If you ran out of time (the BRAKER2 job takes substantial time), you may copy the most important files as follows from a notebook cell:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "cp -r BRAKER2_precomputed_results BRAKER2\n",
    "```\n",
    "</details>\n",
    "\n",
    "The most important output files are:\n",
    "\n",
    "   * [BRAKER2/braker.gtf](BRAKER2/braker.gtf)\n",
    "   * [BRAKER2/Augustus/augustus.hints.gtf](BRAKER2/Augustus/augustus.hints.gtf)\n",
    "   * [BRAKER2/GeneMark-EP/genemark.gtf](BRAKER2/GeneMark-EP/genemark.gtf)\n",
    "   * [BRAKER2/hintsfile.gff](BRAKER2/hintsfile.gff)\n",
    "   \n",
    "The file [BRAKER2/what-to-cite.txt](BRAKER2/what-to-cite.txt) advises you on what papers should be cited if you were going to publish a manuscript on a gene set produced with BRAKER2. \n",
    "\n",
    "All methods described for BRAKER3 (BUSCO, number of transcripts, mono:mult exon ratio, etc.) are of course applicable to BRAKER2, GALBA, and BRAKER1, as well. We will skip it here because of time constraints.\n",
    "\n",
    "### GALBA\n",
    "\n",
    "GALBA ([preprint](https://www.biorxiv.org/content/10.1101/2023.04.10.536199v1.abstract), [software](https://github.com/Gaius-Augustus/GALBA)) is a BRAKER-spinoff that uses miniprot ([paper](https://doi.org/10.1093/bioinformatics/btad014), [software](https://github.com/lh3/miniprot)) to generate a training gene set of AUGUSTUS. In contrast to the BRAKER2 and BRAKER3 pipelines, GALBA is not very good at using remotely related protein evidence. However, given reference proteins of several closely related species, GALBA is very good at recovering gene structures, particularly in large vertebrate genomes. You may execute GALBA as follows (using a toy example data set, it executes within a 3 minutes on 4 threads):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ffbc79ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Tue Nov 21 12:36:05 2023: Log information is stored in file /home/jovyan/BRAKER-TSEBRA-Workshop/GALBA/GALBA.log\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "info: Low coverage detected, coverage will be ignored in the high-confidence set.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#*********\n",
      "# WARNING: Number of reliable training genes is low (226). Recommended are at least 600 genes\n",
      "#*********\n",
      "#*********\n",
      "# WARNING: Number of reliable training genes is low (226). Recommended are at least 600 genes\n",
      "#*********\n",
      "# IMPORTANT INFORMATION: the final output files \n",
      "of this GALBA run are galba.gtf, galba.codingseq, and galba.aa\n",
      "These files are exact copies auf augustus.hints predictions.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t2m11.879s\n",
      "user\t2m11.531s\n",
      "sys\t0m7.876s\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "T=6 # adjust to number of threads that you booted with\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d GALBA ]\n",
    "then\n",
    "    rm -rf GALBA\n",
    "fi\n",
    "\n",
    "time galba.pl --workingdir=GALBA --genome=/opt/BRAKER/example/genome.fa \\\n",
    "    --prot_seq=/opt/GALBA/example/proteins.fa \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --skipOptimize \\ # remember to remove both options if you are running a real job\n",
    "    2> galba.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac784782",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Out of time, job died? Click here.</b></summary>\n",
    "It is unlikely that GALBA will note complete, fast. However, you may copy the most important files as follows from a notebook cell:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "cp -r GALBA_precomputed_results GALBA\n",
    "```\n",
    "</details>\n",
    "\n",
    "The most important output files are:\n",
    "\n",
    "   * [GALBA/augustus.hints.gtf](GALBA/augustus.hints.gtf)\n",
    "   * [GALBA/hintsfile.gff](GALBA/hintsfile.gff)\n",
    "   \n",
    "The file [GALBA/what-to-cite.txt](GALBA/what-to-cite.txt) advises you on what papers should be cited if you were going to publish a manuscript on a gene set produced with GALBA."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a300410d-0522-4f4a-9400-88d63eab6e58",
   "metadata": {},
   "source": [
    "### BRAKER1\n",
    "\n",
    "Since BRAKER3, the pipeline for input of both RNA-Seq and a large database of proteins achieves usually higher accuracy than BRAKER1 with RNA-Seq, only, BRAKER1 is now rather a pipeline that we may resort to using if BRAKER3 died due to insufficient data. BRAKER1 also requires a certain amount of RNA-Seq alignments but that is less than what is required for transcriptome assembly with StringTie in BRAKER3.\n",
    "\n",
    "BRAKER1 uses spliced alignment information from RNA-Seq for training GeneMark-ET ([paper](https://doi.org/10.1093/nar/gku557), [software](http://exon.gatech.edu/genemark/license_download.cgi)), for selecting a training gene set for AUGUSTUS, and for predicting genes with AUGUSTUS. \n",
    "\n",
    "We will run BRAKER1 to predict genes in the genomic sequence with the prepared RNA-Seq intron evidence. As before, we introduce options to save runtime (see BRAKER3 and BRAKER2) that should not be applied in a real-life annotation project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f2c34fe6-3fb7-443f-a026-2a8bf77d304e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Tue Nov 21 12:41:00 2023: Log information is stored in file /home/jovyan/BRAKER-TSEBRA-Workshop/BRAKER1/braker.log\n",
      "#*********\n",
      "# WARNING: Number of reliable training genes is low (199). Recommended are at least 600 genes\n",
      "#*********\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t2m1.412s\n",
      "user\t3m51.685s\n",
      "sys\t0m10.232s\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "T=6 # adjust to number of threads that you booted with, takes ~2.5 minutes on 4 threads\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d BRAKER1 ]\n",
    "then\n",
    "    rm -rf BRAKER1\n",
    "fi\n",
    "\n",
    "time braker.pl --workingdir=BRAKER1 --genome=/opt/BRAKER/example/genome.fa \\\n",
    "    --bam=/opt/BRAKER/example/RNAseq.bam --softmasking \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --gm_max_intergenic 10000 --skipOptimize #  remember to remove this option if you are running a real job\n",
    "    # this call takes a few minutes even with --skipOptimize"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cff6a614-a5ca-445c-87d9-6c06614f1c34",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Out of time, job died? Click here.</b></summary>\n",
    "This job should easily complete within a few minutes, but you may copy the most important files as follows from a notebook cell:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "cp -r BRAKER1_precomputed_results BRAKER1\n",
    "```\n",
    "</details>\n",
    "\n",
    "Note that BRAKER by default expects scripts and binaries in a location relative to the `$AUGUSTUS_CONFIG_PATH`. We here changed the location of the `$AUGUSTUS_CONFIG_PATH` to a writable location. Therefore, we have to tell BRAKER where the scripts and binaries are (`--AUGUSTUS_BIN_PATH`, `--AUGUSTUS_CONFIG_PATH`).\n",
    "\n",
    "The most important output files that we will later use for running TSEBRA are \n",
    "\n",
    "   * [BRAKER1/braker.gtf](BRAKER1/braker.gtf)\n",
    "   * [BRAKER1/Augustus/augustus.hints.gtf](BRAKER1/Augustus/augustus.hints.gtf)\n",
    "   * [BRAKER1/GeneMark-EP/genemark.gtf](BRAKER1/GeneMark-EP/genemark.gtf)\n",
    "   * [BRAKER1/hintsfile.gff](BRAKER1/hintsfile.gff)\n",
    "   \n",
    "The file [BRAKER1/what-to-cite.txt](BRAKER1/what-to-cite.txt) advises you on what papers should be cited if you were going to publish a manuscript on a gene set produced with BRAKER1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0308320b-6e94-4e26-aeb4-4cb138592ccd",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Out of time, job died? Click here.</b></summary>\n",
    "This job should easily complete within a few minutes, but you may copy the most important files as follows from a notebook cell:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "cp -r BRAKER1_precomputed_results BRAKER1\n",
    "```\n",
    "</details>\n",
    "\n",
    "Note that BRAKER by default expects scripts and binaries in a location relative to the `$AUGUSTUS_CONFIG_PATH`. We here changed the location of the `$AUGUSTUS_CONFIG_PATH` to a writable location. Therefore, we have to tell BRAKER where the scripts and binaries are (`--AUGUSTUS_BIN_PATH`, `--AUGUSTUS_CONFIG_PATH`).\n",
    "\n",
    "The most important output files that we will later use for running TSEBRA are \n",
    "\n",
    "   * [BRAKER1/braker.gtf](BRAKER1/braker.gtf)\n",
    "   * [BRAKER1/Augustus/augustus.hints.gtf](BRAKER1/Augustus/augustus.hints.gtf)\n",
    "   * [BRAKER1/GeneMark-EP/genemark.gtf](BRAKER1/GeneMark-EP/genemark.gtf)\n",
    "   * [BRAKER1/hintsfile.gff](BRAKER1/hintsfile.gff)\n",
    "   \n",
    "The file [BRAKER1/what-to-cite.txt](BRAKER1/what-to-cite.txt) advises you on what papers should be cited if you were going to publish a manuscript on a gene set produced with BRAKER1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aee3d341-e69d-4ed5-a1ef-5af40a066f74",
   "metadata": {},
   "source": [
    "### TSEBRA\n",
    "\n",
    "TSEBRA is a tool for selecting a highly accurate gene set from several input sets according to supporting extrinsic evidence. BRAKER internally executes TSEBRA to combine the GeneMark and the AUGUSTUS gene set. If all went well, you do not have run TSEBRA, separately, at all. However, there are several scenarios where TSEBRA may be useful:\n",
    "\n",
    "   * you are unhappy with the braker.gtf file of any BRAKER version while the files of AUGSTUS and GeneMark look fine. In this case, the supporting evidence was so weak that too many gene models got discarded during the original TSEBRA run within BRAKER. The solution here is to pick one of the gene sets (e.g. AUGUSTUS if that seems to be the better set)\n",
    "   \n",
    "   * BRAKER3 failed executing because the provided RNA-Seq data set was too small. In this case, you may wish to combine a BRAKER1 and a BRAKER2 gene set.\n",
    "   \n",
    "First, we will have a look at how to generally run TSEBRA on the example of merging the BRAKER1 and BRAKER2 gene set according to the respective evidence of these runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f8600d47-a94f-4797-b30f-f1fcad3f0416",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "\n",
    "mkdir TSEBRA\n",
    "cd TSEBRA\n",
    "tsebra.py -g ../BRAKER1/Augustus/augustus.hints.gtf,../BRAKER1/GeneMark-ET/genemark.gtf,../BRAKER2/Augustus/augustus.hints.gtf,../BRAKER2/GeneMark-EP/genemark.gtf \\\n",
    "    -e ../BRAKER1/hintsfile.gff,../BRAKER2/hintsfile.gff -o tsebra.gtf 2> tsebra.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6513be99-73a5-4d19-9790-0b6ee16c45e9",
   "metadata": {},
   "source": [
    "Check the file [tsebra.log](tsebra.log) for possible errors. The final gene set is in file [tsebra.gtf](tsebra.gtf)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "990e47da",
   "metadata": {},
   "source": [
    "If you applied this to a real full genome dataset, you want to see an improvement in the TSEBRA output compared to all the inputs. One option is to compare BUSCO scores. In a real scenario, with a complete genome, the BUSCO plot should look more like this (sensitivity should increase in the final TSEBRA set):\n",
    "\n",
    "<img src=\"busco_ideally.png\" alt=\"BUSCO results (ideally)\" width=\"600\"/>\n",
    "\n",
    "As a second example, we will assume that the braker.gtf file of a BRAKER3 run was suboptimal. BUSCO scores and other statistics indicated that the AUGUSTUS gene set alone was quite good, but you want to merge the GeneMark-ETP gene set on top of the AUGUSTUS gene set (keeping all transcripts from the AUGUSTUS gene set). We can do it like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6c164298",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "\n",
    "cd BRAKER3\n",
    "tsebra.py -k Augustus/augustus.hints.gtf -g GeneMark-ETP/genemark.gtf \\\n",
    "    -e hintsfile.gff -o tsebra.gtf 2> tsebra.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "022a551a-5464-4450-9fee-f910798fd844",
   "metadata": {},
   "source": [
    "## Data visualization in the UCSC Genome Browser\n",
    "\n",
    "Visualization of gene structures in context with extrinsic evidence is essential for coming to a decision on whether a gene set \"makes sense\" or \"does not make sense\". Typical problems that you may observe in a genome browser include \"split genes\" (where evidence implies two genes should in fact be a single gene) or \"joined genes\" (where evidence implies one gene should be split into two genes).\n",
    "\n",
    "The UCSC Genome Browser ([paper](https://doi.org/10.1101/gr.229102), [resource](https://genome.ucsc.edu/)) is one of the most popular genome browsers. It has the advantage that you do not have to install a browser instance on your own webserver. Instead, you only need to provide a certain data structure with your target data on a webserver. The UCSC Genome Browser servers can display your data from there. The data structures are called \"track data hubs\" or \"assembly hubs\" ([paper](https://doi.org/10.1093/bioinformatics/btt637)). \n",
    "\n",
    "MakeHub ([paper](https://doi.org/10.1016/j.gpb.2019.05.003), [software](https://github.com/Gaius-Augustus/MakeHub )) is a python script that fully automates the generation of such track data hubs for novel genomes. In the following, we will generate a simple track data hub for the genome sequence that we annotated with BRAKER3 (takes only a few seconds):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "65c6f7a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Was unable to locate file bigGenePred.as in /opt/MakeHub and ./tmp-at_chunk/, will try to download it from http://genome.ucsc.edu/goldenPath/help/examples/...\n",
      "Trying to download http://genome.ucsc.edu/goldenPath/help/examples/bigGenePred.as and store into ./tmp-at_chunk//bigGenePred.as\n",
      "Stored file at ./tmp-at_chunk//bigGenePred.as.\n",
      "Was unable to locate file cytoBand.as in /opt/MakeHub and ./tmp-at_chunk/, will try to download it from http://genome.ucsc.edu/goldenPath/help/examples/...\n",
      "Trying to download http://genome.ucsc.edu/goldenPath/help/examples/cytoBand.as and store into ./tmp-at_chunk//cytoBand.as\n",
      "Stored file at ./tmp-at_chunk//cytoBand.as.\n",
      "Generating genome 2bit file ./at_chunk/at_chunk/at_chunk.2bit...\n",
      "Generating chromsome size info file ./at_chunk/at_chunk/at_chunk.chrom.sizes...\n",
      "Generating variable step wiggle file for GC content ./tmp-at_chunk/at_chunk.gc5Base.wigVarStep.gz...\n",
      "Generating bigWig file for GC content ./at_chunk/at_chunk/at_chunk.gc5Base.bw...\n",
      "Generating cytoband track...\n",
      "Generating bigBed file for ./tmp-at_chunk/at_chunk.cytoband.bed...\n",
      "Generating softmasking information bed3 file ./tmp-at_chunk/at_chunk.RMsoft.bed3 from genome data (this may take a while)...\n",
      "Sorting file ./tmp-at_chunk/at_chunk.RMsoft.bed3...\n",
      "Generating bigBed file for ./tmp-at_chunk/at_chunk.RMsoft.s.bed3...\n",
      "Generating bigWig RNA-Seq track(s) from BAM...\n",
      "Converting compressed wig to bigWig\n",
      "Writing to track db\n",
      "Generating gene track from file BRAKER3/braker.gtf with label BRAKER3\n",
      "Generating bigBed file from bigGenePred format for ./tmp-at_chunk/BRAKER3.fixed.bed...\n",
      "Next step should be writing trackDb entry...\n",
      "Separating hints into separate tracks...\n",
      "Generating bigBed file for ./tmp-at_chunk/b2h_intron.hints_1.sorted...\n",
      "Generating bigBed file for ./tmp-at_chunk/ProtHint_start.hints_1.sorted...\n",
      "Generating bigBed file for ./tmp-at_chunk/ProtHint_intron.hints_1.sorted...\n",
      "Generating bigBed file for ./tmp-at_chunk/ProtHint_stop.hints_1.sorted...\n",
      "Generating bigBed file for ./tmp-at_chunk/ProtHint_CDSpart.hints_1.bed...\n",
      "\n",
      "Hub is ready, please copy to a web server, e.g.\n",
      "\"scp -r ./at_chunk user@server:/target/location\"\n",
      "Feed the link to hub.txt to UCSC genome browser:\n",
      "\tMy Data -> Track Hubs -> MyHubs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t0m9.281s\n",
      "user\t0m14.847s\n",
      "sys\t0m2.947s\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "T=6 # adjust to number of threads that you booted with\n",
    "\n",
    "time make_hub.py -e katharina.hoff@uni-greifswald.de \\\n",
    "    --genome /opt/BRAKER/example/genome.fa --long_label \"A chunk from the Arabidopsis thaliana genome\" \\\n",
    "    --short_label at_chunk  --bam /opt/BRAKER/example/RNAseq.bam --threads ${T} \\\n",
    "    --latin_name \"Arabidopsis thaliana\" \\\n",
    "    --assembly_version \"artifically split custom assembly\" \\\n",
    "    --hints BRAKER3/hintsfile.gff --gene_track BRAKER3/braker.gtf BRAKER3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc653250-d5b7-4714-a346-e22563d67bc9",
   "metadata": {},
   "source": [
    "You can't perform the suggested `scp` command from the apphub, unless you have privileges on a University of Greifswald webserver. We have therefore copied a prepared hub in advance. The `hub.txt` is available at https://bioinf.uni-greifswald.de/hubs/at_chunk/hub.txt . Remember that link.\n",
    "\n",
    "In order to visualize your data, go to https://genome.ucsc.edu/ . Click on `My Data` -> `Track Hubs` -> choose the European mirror -> click on `Connected Hubs` and enter the link https://bioinf.uni-greifswald.de/hubs/at_chunk/hub.txt into the text window -> click on `Add Hub`. Congratulations, your Hub is now connected. You should be able to browse something like this: \n",
    "\n",
    "<img src=\"at_chunk.png\" alt=\"UCSC Genome Browser example\" width=\"1000\"/>\n",
    "\n",
    "### How to know which sequences to browse\n",
    "\n",
    "The long sequences are usually the most interesting to look at. The following command gives you the names of sequences in the order of descending length, you can copy-paste the sequence names into the search window in the UCSC Genome Browser."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "72bd61f2-9007-45bd-83bf-5af23fee0995",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X1\t125040\n",
      "X2\t125100\n",
      "X3\t125100\n",
      "X4\t125100\n",
      "X5\t125100\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "\n",
    "N=5 # how many longest sequences would you like to know about\n",
    "\n",
    "summarizeACGTcontent.pl /opt/BRAKER/example/genome.fa | grep bases | head -${N} | sort -n \\\n",
    "   | perl -ne 'm/(\\d+)\\s+bases\\.\\s+(\\S+)/; print \"$2\\t$1\\n\";'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7681d88-0284-4819-bf10-955279525239",
   "metadata": {},
   "source": [
    "## How to run BRAKER, GALBA, TSEBRA, and other software in Docker\n",
    "\n",
    "If you have a machine on which you have root permissions and Docker, you can run the exact same container as we have been using during this workshop as follows:\n",
    "\n",
    "```\n",
    "sudo docker run --rm -it -u root katharinahoff/bioinformatics-notebook:latest bash\n",
    "```\n",
    "\n",
    "You can execute all shell commands that we covered in this notebook in that container.\n",
    "\n",
    "If you want to render the JuypterNotebook, start the docker container like this:\n",
    "\n",
    "```\n",
    "sudo docker run --rm -it -u 1000:0 -p 8888:8888 katharinahoff/bioinformatics-notebook:latest\n",
    "```\n",
    "\n",
    "and open the link that appears in Chrome Browser.\n",
    "\n",
    "## How to run BRAKER, GALBA, TSEBRA, and other software in Singularity\n",
    "\n",
    "If you want to use Singularity, you first need to build an image file:\n",
    "\n",
    "```\n",
    "singularity build bioinformatics.sif katharinahoff/bioinformatics-notebook:latest\n",
    "\n",
    "```\n",
    "\n",
    "Using Singularity, you can run the exact same container as we have been using during this workshop with JupyterNotebook rendering as follows:\n",
    "\n",
    "```\n",
    "singularity exec --cleanenv --bind ${PWD}:${PWD} --bind $PWD:/home/jovyan bioinformatics.sif jupyter notebook --no-browser --ip=127.0.0.1\n",
    "```\n",
    "\n",
    "(You possibly need to mount more directories with bind in order to access your raw data in a real use case.)\n",
    "\n",
    "## Troubleshooting\n",
    "\n",
    "### I have 80,000 genes predicted by BRAKER/TSEBRA in a full genome, what shall I do?\n",
    "\n",
    "Please first check whether you are referring to genes, or to transcripts. BRAKER predicts alternative isoforms. If RNA-Seq data supports this, the number of alternative transcripts may be large, but likely true. If it's really genes that you counted, then 80,000 sounds way too much, indeed (unless you are dealing with a genome that has multiple copies of each chromosome). Most likely, GeneMark-ET/ES/EP/ETP produced highly fragmented training genes for AUGUSTUS. This will also lead to highly fragmented genes predicted by AUGUSTUS. First, check whether your genome has been masked for repeats. Consider using the additional TRF masking desribed at the top of this notebook. If that does not help, and if you have a protein set of closely related species at hand, consider using that protein set as sole training data for AUGUSTUS. You can use GALBA for this (https://github.com/Gaius-Augustus/GALBA).\n",
    "\n",
    "### I have only 10,000 genes predicted by BRAKER/TSEBRA in a full genome, what shall I do?\n",
    "\n",
    "First: be sure that you expect more than 10,000 genes. There are parasites with few genes. Check whether the BRAKER output files in subfolders Augustus and GeneMark-* produced more genes than TSEBRA. By default, TSEBRA will discard genes without evidence. If you have only little evidence for your species, TSEBRA might be a bad idea. You can also try rerunning TSEBRA with enforcing one of the gene sets.  There are also species for which is \"normal\" to observe less than 10,000 genes, check the annotated relatives.\n",
    "\n",
    "### How do I know how many genes to expect?\n",
    "\n",
    "Hard to say. You can download gene sets of related species e.g. from NCBI Genomes, and count. Some gene sets tend to be \"underannotated\", i.e. they may represent rather the lower numbers of what might be realistic. Katharina usually gets nervous about more than 45,000 genes and fewer than 15,000 genes. These are definitely weird gene counts - but as stated before, there are cases where these are totally fine, too. Otherwise: always inspect your gene set in a Genome Browser such as the UCSC Genome Browser to identify problems.\n",
    "\n",
    "### I have long isoseq RNA-Seq transcripts, can I put them into BRAKER?\n",
    "\n",
    "No. But we have [other instructions](https://github.com/Gaius-Augustus/BRAKER/blob/master/docs/long_reads/long_read_protocol.md) for you. Please note: isoseq data does not always aid structural genome annotation over short read RNA-Seq data.\n",
    "\n",
    "### The BUSCO scores of my genome are higher than those of my protein, why?\n",
    "\n",
    "Some short housekeeping genes are commonly masked during repeat masking. They won't show up in a BRAKER gene set. You can manually add them from BUSCO output to a final gene set. Beware: not every genome-level reported BUSCO has a valid gene structure. We currently do not provide an automated workflow for this.\n",
    "\n",
    "### I opened an issue on GitHub about BRAKER or TSEBRA 100 days ago, nobody replied, why?\n",
    "\n",
    "We are a small team of developers. We try our best and usually respond to well described and easy-to-solve issues within a rather short time frame. Solving other issues may take considerable amounts of time that we simply do not have, or issues may be described in a way that we don't know what do with them... please be patient with us.\n",
    "\n",
    "### I have a problem, whom do I tell?\n",
    "\n",
    "Please read through the Issues on Github. If the issue does not exist, yet, open an issue. Do not send e-mails. We will likely reply: please open an issue on GitHub.\n",
    "\n",
    "# Ready to move on?\n",
    "\n",
    "If you feel confident about your skills, take them to the next level. We have prepared chromosome 4 of a small genome in the following notebook: [Annotate_Babesia_duncani.ipynb](Annotate_Babesia_duncani.ipynb). Note that you may be missing data if you try to work on this exercise in Docker, locally. You can download 1.8 GB to your machine if you like, a script for this is mentioned in the Notebook.\n",
    "\n",
    "### The End"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc60ef6b-a759-4501-bbab-bf3732c18324",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
